diff --git a/diagrams/quilt.cpp.dia b/diagrams/quilt.cpp.dia
index 86b113c..fe8c45a 100644
--- a/diagrams/quilt.cpp.dia
+++ b/diagrams/quilt.cpp.dia
@@ -2090,13 +2090,13 @@ seems like the least bad approach.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O16">
       <dia:attribute name="obj_pos">
-        <dia:point val="388,200"/>
+        <dia:point val="388,201"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="387.95,199.95;399.395,203.05"/>
+        <dia:rectangle val="387.95,200.95;399.395,204.05"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="388,200"/>
+        <dia:point val="388,201"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="11.345000000000001"/>
@@ -2684,22 +2684,22 @@ seems like the least bad approach.
     </dia:object>
     <dia:object type="UML - LargePackage" version="0" id="O22">
       <dia:attribute name="obj_pos">
-        <dia:point val="480,213.854"/>
+        <dia:point val="480,187"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="479.95,212.804;525.05,237.904"/>
+        <dia:rectangle val="479.95,185.95;525.05,238.05"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="480,213.854"/>
+        <dia:point val="480,187"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="45"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
-        <dia:real val="24"/>
+        <dia:real val="51"/>
       </dia:attribute>
       <dia:attribute name="line_width">
         <dia:real val="0.10000000149011612"/>
@@ -2723,13 +2723,13 @@ seems like the least bad approach.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O23">
       <dia:attribute name="obj_pos">
-        <dia:point val="482,221.854"/>
+        <dia:point val="482,195"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="481.95,221.804;493.12,224.904"/>
+        <dia:rectangle val="481.95,194.95;493.12,198.05"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="482,221.854"/>
+        <dia:point val="482,195"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="11.07"/>
@@ -2834,13 +2834,13 @@ seems like the least bad approach.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O24">
       <dia:attribute name="obj_pos">
-        <dia:point val="503,221.854"/>
+        <dia:point val="503,195"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="502.95,221.804;518.1,224.904"/>
+        <dia:rectangle val="502.95,194.95;518.1,198.05"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="503,221.854"/>
+        <dia:point val="503,195"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="15.050000000000001"/>
@@ -4355,16 +4355,16 @@ C++ language.
     </dia:object>
     <dia:object type="UML - Note" version="0" id="O42">
       <dia:attribute name="obj_pos">
-        <dia:point val="482,215.854"/>
+        <dia:point val="482,189"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="481.95,215.804;497.58,220.004"/>
+        <dia:rectangle val="481.95,188.95;497.58,193.15"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="482,215.854"/>
+        <dia:point val="482,189"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="15.530000000000001"/>
@@ -4396,7 +4396,7 @@ facet of the C++ language.
             <dia:real val="0.80000000000000004"/>
           </dia:attribute>
           <dia:attribute name="pos">
-            <dia:point val="482.35,217.099"/>
+            <dia:point val="482.35,190.245"/>
           </dia:attribute>
           <dia:attribute name="color">
             <dia:color val="#000000"/>
@@ -5337,13 +5337,13 @@ facet of the C++ language.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O49">
       <dia:attribute name="obj_pos">
-        <dia:point val="517,216"/>
+        <dia:point val="517,189.146"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="516.95,215.95;521.785,219.05"/>
+        <dia:rectangle val="516.95,189.096;521.785,192.196"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="517,216"/>
+        <dia:point val="517,189.146"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="4.7350000000000003"/>
@@ -9979,7 +9979,7 @@ registry of all available formatters.#</dia:string>
         <dia:point val="260.762,153.4"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="248.602,152.65;261.512,163.8"/>
+        <dia:rectangle val="248.603,152.65;261.512,163.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -10661,7 +10661,7 @@ registry of all available formatters.#</dia:string>
         <dia:point val="169.81,198"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="131.477,192.25;170.56,198.8"/>
+        <dia:rectangle val="131.478,192.25;170.56,198.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -11850,13 +11850,13 @@ main, or from another initialiser.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O107">
       <dia:attribute name="obj_pos">
-        <dia:point val="503,216"/>
+        <dia:point val="503,189.146"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="502.95,215.95;508.555,219.65"/>
+        <dia:rectangle val="502.95,189.096;508.555,192.796"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="503,216"/>
+        <dia:point val="503,189.146"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="5.5049999999999999"/>
@@ -14020,7 +14020,7 @@ main, or from another initialiser.#</dia:string>
         <dia:point val="355,-46"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="354.25,-54.6666;445.822,-28.2"/>
+        <dia:rectangle val="354.25,-54.6666;445.823,-28.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -15919,13 +15919,13 @@ main, or from another initialiser.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O128">
       <dia:attribute name="obj_pos">
-        <dia:point val="482,232"/>
+        <dia:point val="482,205.146"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="481.95,231.95;496.85,235.05"/>
+        <dia:rectangle val="481.95,205.096;496.85,208.196"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="482,232"/>
+        <dia:point val="482,205.146"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="14.800000000000001"/>
@@ -19897,7 +19897,7 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
         <dia:point val="250,-9.6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="245.412,-10.35;250.75,3.74958"/>
+        <dia:rectangle val="245.413,-10.35;250.75,3.74958"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -21737,7 +21737,7 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
         <dia:point val="262,-41.9166"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="252.947,-42.6666;262.75,-28.2"/>
+        <dia:rectangle val="252.948,-42.6666;262.75,-28.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -22859,13 +22859,13 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O185">
       <dia:attribute name="obj_pos">
-        <dia:point val="481,186"/>
+        <dia:point val="530,221"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="480.95,185.95;527.365,199.05"/>
+        <dia:rectangle val="529.95,220.95;576.365,234.05"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="481,186"/>
+        <dia:point val="530,221"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="46.314999999999998"/>
@@ -23459,13 +23459,13 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O186">
       <dia:attribute name="obj_pos">
-        <dia:point val="482,227"/>
+        <dia:point val="482,200.146"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="481.95,226.95;494.853,230.05"/>
+        <dia:rectangle val="481.95,200.096;494.853,203.196"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="482,227"/>
+        <dia:point val="482,200.146"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="12.8025"/>
@@ -23570,13 +23570,13 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O187">
       <dia:attribute name="obj_pos">
-        <dia:point val="503,227"/>
+        <dia:point val="503,200.146"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="502.95,226.95;519.832,230.05"/>
+        <dia:rectangle val="502.95,200.096;519.832,203.196"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="503,227"/>
+        <dia:point val="503,200.146"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="16.782499999999999"/>
@@ -23681,13 +23681,13 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O188">
       <dia:attribute name="obj_pos">
-        <dia:point val="503,232"/>
+        <dia:point val="503,205.146"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="502.95,231.95;514.395,235.05"/>
+        <dia:rectangle val="502.95,205.096;514.395,208.196"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="503,232"/>
+        <dia:point val="503,205.146"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="11.345000000000001"/>
@@ -28856,7 +28856,7 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
         <dia:point val="320.962,80"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="320.212,75.25;340.502,80.8"/>
+        <dia:rectangle val="320.212,75.25;340.503,80.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -29885,7 +29885,7 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
         <dia:point val="181.565,163"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="180.815,156.25;212.887,163.8"/>
+        <dia:rectangle val="180.815,156.25;212.888,163.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -30029,7 +30029,7 @@ Note that the facet directory may be empty if the user disabled facet folders.#<
         <dia:point val="205.993,178"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="205.243,165.75;212.887,178.8"/>
+        <dia:rectangle val="205.243,165.75;212.888,178.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -32285,7 +32285,7 @@ an enumeration, false otherwise.#</dia:string>
         <dia:point val="546.745,22.6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="545.995,21.85;569.043,34.8"/>
+        <dia:rectangle val="545.995,21.85;569.042,34.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -34561,7 +34561,7 @@ and element.#</dia:string>
         <dia:point val="374.805,151"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="347.624,144.95;375.555,152.6"/>
+        <dia:rectangle val="347.624,144.25;375.555,151.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -34592,5 +34592,671 @@ and element.#</dia:string>
       </dia:connections>
       <dia:childnode parent="O3"/>
     </dia:object>
+    <dia:object type="UML - Class" version="0" id="O278">
+      <dia:attribute name="obj_pos">
+        <dia:point val="363,213"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="362.95,212.95;375.952,216.05"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="363,213"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="12.9025"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="3"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#primitive_header_formatter#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##DOGEN yarn.generalization.parent=formatters::file_formatter_interface#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations"/>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+      <dia:childnode parent="O4"/>
+    </dia:object>
+    <dia:object type="UML - Class" version="0" id="O279">
+      <dia:attribute name="obj_pos">
+        <dia:point val="408,201"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="407.95,200.95;420.952,204.05"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="408,201"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="12.9025"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="3"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#primitive_header_formatter#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##DOGEN yarn.generalization.parent=formatters::file_formatter_interface#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations"/>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+      <dia:childnode parent="O8"/>
+    </dia:object>
+    <dia:object type="UML - Class" version="0" id="O280">
+      <dia:attribute name="obj_pos">
+        <dia:point val="388,232"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="387.95,231.95;400.952,235.05"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="388,232"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="12.9025"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="3"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#primitive_header_formatter#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##DOGEN yarn.generalization.parent=formatters::file_formatter_interface#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations"/>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+      <dia:childnode parent="O26"/>
+    </dia:object>
+    <dia:object type="UML - Class" version="0" id="O281">
+      <dia:attribute name="obj_pos">
+        <dia:point val="434,205"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="433.95,204.95;446.952,208.05"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="434,205"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="12.9025"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="3"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#primitive_header_formatter#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##DOGEN yarn.generalization.parent=formatters::file_formatter_interface#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations"/>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+      <dia:childnode parent="O17"/>
+    </dia:object>
+    <dia:object type="UML - Class" version="0" id="O282">
+      <dia:attribute name="obj_pos">
+        <dia:point val="454,227"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="453.95,226.95;466.952,230.05"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="454,227"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="12.9025"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="3"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#primitive_header_formatter#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##DOGEN yarn.generalization.parent=formatters::file_formatter_interface#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations"/>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+      <dia:childnode parent="O31"/>
+    </dia:object>
+    <dia:object type="UML - Class" version="0" id="O283">
+      <dia:attribute name="obj_pos">
+        <dia:point val="482,210"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="481.95,209.95;494.952,213.05"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="482,210"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="12.9025"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="3"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#primitive_header_formatter#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##DOGEN yarn.generalization.parent=formatters::file_formatter_interface#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations"/>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+      <dia:childnode parent="O22"/>
+    </dia:object>
   </dia:layer>
 </dia:diagram>
diff --git a/doc/agile/sprint_backlog_88.org b/doc/agile/sprint_backlog_88.org
index ac52f60..86f9f3b 100644
--- a/doc/agile/sprint_backlog_88.org
+++ b/doc/agile/sprint_backlog_88.org
@@ -1043,6 +1043,8 @@ It seems its fairly trivial to extend dia using Pything:
 - [[https://github.com/GNOME/dia/tree/master/plug-ins/python][Lots of examples of python scripts]]
 
 *** STARTED Consider merging provider interface with file formatter   :story:
+    CLOCK: [2016-09-23 Fri 14:12]--[2016-09-23 Fri 15:39] =>  1:27
+    CLOCK: [2016-09-23 Fri 13:17]--[2016-09-23 Fri 13:41] =>  0:24
     CLOCK: [2016-09-23 Fri 13:07]--[2016-09-23 Fri 13:16] =>  0:09
     CLOCK: [2016-09-23 Fri 11:52]--[2016-09-23 Fri 12:29] =>  0:37
     CLOCK: [2016-09-23 Fri 10:21]--[2016-09-23 Fri 11:51] =>  1:30
@@ -1293,6 +1295,47 @@ Other names:
 - canonical formatter
 - reference formatter
 
+We need to support a strange use case: where the formatter does not
+exist for a given element type. For example, we do not have primitive
+formatters, but there are directives set in them:
+
+#+begin_src json-mode
+        {
+            "meta_type" : "primitive",
+            "simple_name" : "uint64_t",
+            "extensions" : {
+                "quilt.cpp.helper.family" : "Number",
+                "quilt.cpp.aspect.requires_manual_default_constructor" : true,
+                "quilt.cpp.types.class_header_formatter.inclusion_directive" : "<cstdint>",
+                "quilt.cpp.hash.class_header_formatter.inclusion_required" : false,
+                "quilt.cpp.io.class_header_formatter.inclusion_required" : false,
+                "quilt.cpp.test_data.class_header_formatter.inclusion_required" : false,
+                "quilt.cpp.serialization.class_header_formatter.inclusion_required" : false,
+                "quilt.cpp.odb.class_header_formatter.inclusion_required" : false
+            }
+        },
+#+end_src
+
+The problem with this is that if we do not have a formatter for
+primitives, then we will not read the directives. In the past this
+worked because we were processing the cross-product of formatters and
+element sub-types, so the mistake of
+=quilt.cpp.types.class_header_formatter.inclusion_directive= was
+actually resulted in the correct result. But of course, we cannot
+replace class_header_formatter with the correct formatter name (as we
+don't have one). Nor does it sound good to have to hard-code the
+formatter name against the type. One way to solve this is with
+canonical formatters:
+
+- use the canonical formatter name in the declaration
+- ensure we always read directives for the canonical formatter from
+  the meta-data.
+- when processing, only set the canonical formatter if it was not
+  already set by meta-data.
+
+When testing the fix, we need to delete the mock formaters created for
+primitives.
+
 *** Tidy-up of inclusion terminology                                  :story:
 
 Random notes:
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/all.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/all.hpp
index 2b7b935..c7541d4 100644
--- a/projects/quilt.cpp/include/dogen/quilt.cpp/types/all.hpp
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/all.hpp
@@ -120,6 +120,7 @@
 #include "dogen/quilt.cpp/types/formatters/include_cmakelists_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/types/class_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formattables/path_derivatives_repository.hpp"
+#include "dogen/quilt.cpp/types/formatters/io/primitive_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/aspect_annotations_repository.hpp"
 #include "dogen/quilt.cpp/types/annotations/helper_annotations_repository.hpp"
 #include "dogen/quilt.cpp/types/annotations/streaming_annotations_factory.hpp"
@@ -127,10 +128,12 @@
 #include "dogen/quilt.cpp/types/formattables/formatter_properties_factory.hpp"
 #include "dogen/quilt.cpp/types/formattables/global_enablement_properties.hpp"
 #include "dogen/quilt.cpp/types/formattables/helper_properties_repository.hpp"
+#include "dogen/quilt.cpp/types/formatters/odb/primitive_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/types/visitor_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/element_annotations_repository.hpp"
 #include "dogen/quilt.cpp/types/formattables/element_properties_repository.hpp"
 #include "dogen/quilt.cpp/types/formattables/enablement_repository_factory.hpp"
+#include "dogen/quilt.cpp/types/formatters/hash/primitive_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/test_data/enum_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/inclusion_directive_annotations.hpp"
 #include "dogen/quilt.cpp/types/formattables/inclusion_dependencies_builder.hpp"
@@ -139,6 +142,7 @@
 #include "dogen/quilt.cpp/types/formatters/test_data/class_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/types/exception_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/types/namespace_header_formatter.hpp"
+#include "dogen/quilt.cpp/types/formatters/types/primitive_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/streaming_annotations_repository.hpp"
 #include "dogen/quilt.cpp/types/formattables/formatter_properties_repository.hpp"
 #include "dogen/quilt.cpp/types/formattables/inclusion_directives_repository.hpp"
@@ -148,6 +152,7 @@
 #include "dogen/quilt.cpp/types/formatters/serialization/enum_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/directory_names_annotations_factory.hpp"
 #include "dogen/quilt.cpp/types/formatters/serialization/class_header_formatter.hpp"
+#include "dogen/quilt.cpp/types/formatters/test_data/primitive_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/types/class_implementation_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/types/forward_declarations_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/opaque_annotations_factory_interface.hpp"
@@ -161,6 +166,7 @@
 #include "dogen/quilt.cpp/types/formatters/test_data/enum_implementation_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/inclusion_directive_annotations_factory.hpp"
 #include "dogen/quilt.cpp/types/formattables/inclusion_dependencies_builder_factory.hpp"
+#include "dogen/quilt.cpp/types/formatters/serialization/primitive_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/serialization/registrar_header_formatter.hpp"
 #include "dogen/quilt.cpp/types/formatters/test_data/class_implementation_formatter.hpp"
 #include "dogen/quilt.cpp/types/annotations/streaming_annotations_repository_factory.hpp"
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/hash/primitive_header_formatter.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/hash/primitive_header_formatter.hpp
new file mode 100644
index 0000000..4e89bd5
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/hash/primitive_header_formatter.hpp
@@ -0,0 +1,70 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_HASH_PRIMITIVE_HEADER_FORMATTER_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_HASH_PRIMITIVE_HEADER_FORMATTER_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include "dogen/quilt.cpp/types/formatters/file_formatter_interface.hpp"
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace hash {
+
+class primitive_header_formatter final : public file_formatter_interface {
+public:
+    /**
+     * @brief Returns the formatter name.
+     */
+    static std::string static_formatter_name();
+
+public:
+    std::string id() const override;
+    dynamic::ownership_hierarchy ownership_hierarchy() const override;
+    std::type_index element_type_index() const override;
+
+public:
+    std::list<std::string> inclusion_dependencies(
+        const formattables::inclusion_dependencies_builder_factory& f,
+        const yarn::element& e) const override;
+
+    inclusion_support_types inclusion_support_type() const override;
+
+    boost::filesystem::path inclusion_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+    boost::filesystem::path full_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+public:
+    void register_provider(formattables::registrar& rg) const override;
+
+    dogen::formatters::file
+    format(const context& ctx, const yarn::element& e) const override;
+};
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/hash/primitive_header_formatter_fwd.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/hash/primitive_header_formatter_fwd.hpp
new file mode 100644
index 0000000..befb533
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/hash/primitive_header_formatter_fwd.hpp
@@ -0,0 +1,38 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_HASH_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_HASH_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace hash {
+
+class primitive_header_formatter;
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/primitive_header_formatter.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/primitive_header_formatter.hpp
new file mode 100644
index 0000000..5875fa5
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/primitive_header_formatter.hpp
@@ -0,0 +1,70 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_IO_PRIMITIVE_HEADER_FORMATTER_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_IO_PRIMITIVE_HEADER_FORMATTER_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include "dogen/quilt.cpp/types/formatters/file_formatter_interface.hpp"
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace io {
+
+class primitive_header_formatter final : public file_formatter_interface {
+public:
+    /**
+     * @brief Returns the formatter name.
+     */
+    static std::string static_formatter_name();
+
+public:
+    std::string id() const override;
+    dynamic::ownership_hierarchy ownership_hierarchy() const override;
+    std::type_index element_type_index() const override;
+
+public:
+    std::list<std::string> inclusion_dependencies(
+        const formattables::inclusion_dependencies_builder_factory& f,
+        const yarn::element& e) const override;
+
+    inclusion_support_types inclusion_support_type() const override;
+
+    boost::filesystem::path inclusion_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+    boost::filesystem::path full_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+public:
+    void register_provider(formattables::registrar& rg) const override;
+
+    dogen::formatters::file
+    format(const context& ctx, const yarn::element& e) const override;
+};
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/primitive_header_formatter_fwd.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/primitive_header_formatter_fwd.hpp
new file mode 100644
index 0000000..d17c230
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/primitive_header_formatter_fwd.hpp
@@ -0,0 +1,38 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_IO_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_IO_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace io {
+
+class primitive_header_formatter;
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/traits.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/traits.hpp
index cde5b37..a71a542 100644
--- a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/traits.hpp
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/io/traits.hpp
@@ -58,6 +58,11 @@ struct traits {
      * @brief Name of the enumeration implementation formatter.
      */
     static std::string enum_implementation_formatter_name();
+
+    /**
+     * @brief Name of the primitive implementation formatter.
+     */
+    static std::string primitive_header_formatter_name();
 };
 
 } } } } }
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/odb/primitive_header_formatter.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/odb/primitive_header_formatter.hpp
new file mode 100644
index 0000000..7332310
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/odb/primitive_header_formatter.hpp
@@ -0,0 +1,70 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_ODB_PRIMITIVE_HEADER_FORMATTER_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_ODB_PRIMITIVE_HEADER_FORMATTER_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include "dogen/quilt.cpp/types/formatters/file_formatter_interface.hpp"
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace odb {
+
+class primitive_header_formatter final : public file_formatter_interface {
+public:
+    /**
+     * @brief Returns the formatter name.
+     */
+    static std::string static_formatter_name();
+
+public:
+    std::string id() const override;
+    dynamic::ownership_hierarchy ownership_hierarchy() const override;
+    std::type_index element_type_index() const override;
+
+public:
+    std::list<std::string> inclusion_dependencies(
+        const formattables::inclusion_dependencies_builder_factory& f,
+        const yarn::element& e) const override;
+
+    inclusion_support_types inclusion_support_type() const override;
+
+    boost::filesystem::path inclusion_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+    boost::filesystem::path full_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+public:
+    void register_provider(formattables::registrar& rg) const override;
+
+    dogen::formatters::file
+    format(const context& ctx, const yarn::element& e) const override;
+};
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/odb/primitive_header_formatter_fwd.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/odb/primitive_header_formatter_fwd.hpp
new file mode 100644
index 0000000..ef1078d
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/odb/primitive_header_formatter_fwd.hpp
@@ -0,0 +1,38 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_ODB_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_ODB_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace odb {
+
+class primitive_header_formatter;
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/serialization/primitive_header_formatter.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/serialization/primitive_header_formatter.hpp
new file mode 100644
index 0000000..138161c
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/serialization/primitive_header_formatter.hpp
@@ -0,0 +1,70 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_SERIALIZATION_PRIMITIVE_HEADER_FORMATTER_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_SERIALIZATION_PRIMITIVE_HEADER_FORMATTER_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include "dogen/quilt.cpp/types/formatters/file_formatter_interface.hpp"
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace serialization {
+
+class primitive_header_formatter final : public file_formatter_interface {
+public:
+    /**
+     * @brief Returns the formatter name.
+     */
+    static std::string static_formatter_name();
+
+public:
+    std::string id() const override;
+    dynamic::ownership_hierarchy ownership_hierarchy() const override;
+    std::type_index element_type_index() const override;
+
+public:
+    std::list<std::string> inclusion_dependencies(
+        const formattables::inclusion_dependencies_builder_factory& f,
+        const yarn::element& e) const override;
+
+    inclusion_support_types inclusion_support_type() const override;
+
+    boost::filesystem::path inclusion_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+    boost::filesystem::path full_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+public:
+    void register_provider(formattables::registrar& rg) const override;
+
+    dogen::formatters::file
+    format(const context& ctx, const yarn::element& e) const override;
+};
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/serialization/primitive_header_formatter_fwd.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/serialization/primitive_header_formatter_fwd.hpp
new file mode 100644
index 0000000..e0c4dfc
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/serialization/primitive_header_formatter_fwd.hpp
@@ -0,0 +1,38 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_SERIALIZATION_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_SERIALIZATION_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace serialization {
+
+class primitive_header_formatter;
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/test_data/primitive_header_formatter.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/test_data/primitive_header_formatter.hpp
new file mode 100644
index 0000000..8313d64
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/test_data/primitive_header_formatter.hpp
@@ -0,0 +1,70 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_TEST_DATA_PRIMITIVE_HEADER_FORMATTER_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_TEST_DATA_PRIMITIVE_HEADER_FORMATTER_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include "dogen/quilt.cpp/types/formatters/file_formatter_interface.hpp"
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace test_data {
+
+class primitive_header_formatter final : public file_formatter_interface {
+public:
+    /**
+     * @brief Returns the formatter name.
+     */
+    static std::string static_formatter_name();
+
+public:
+    std::string id() const override;
+    dynamic::ownership_hierarchy ownership_hierarchy() const override;
+    std::type_index element_type_index() const override;
+
+public:
+    std::list<std::string> inclusion_dependencies(
+        const formattables::inclusion_dependencies_builder_factory& f,
+        const yarn::element& e) const override;
+
+    inclusion_support_types inclusion_support_type() const override;
+
+    boost::filesystem::path inclusion_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+    boost::filesystem::path full_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+public:
+    void register_provider(formattables::registrar& rg) const override;
+
+    dogen::formatters::file
+    format(const context& ctx, const yarn::element& e) const override;
+};
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/test_data/primitive_header_formatter_fwd.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/test_data/primitive_header_formatter_fwd.hpp
new file mode 100644
index 0000000..6fad1b3
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/test_data/primitive_header_formatter_fwd.hpp
@@ -0,0 +1,38 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_TEST_DATA_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_TEST_DATA_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace test_data {
+
+class primitive_header_formatter;
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/types/primitive_header_formatter.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/types/primitive_header_formatter.hpp
new file mode 100644
index 0000000..9027a76
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/types/primitive_header_formatter.hpp
@@ -0,0 +1,70 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_TYPES_PRIMITIVE_HEADER_FORMATTER_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_TYPES_PRIMITIVE_HEADER_FORMATTER_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include "dogen/quilt.cpp/types/formatters/file_formatter_interface.hpp"
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace types {
+
+class primitive_header_formatter final : public file_formatter_interface {
+public:
+    /**
+     * @brief Returns the formatter name.
+     */
+    static std::string static_formatter_name();
+
+public:
+    std::string id() const override;
+    dynamic::ownership_hierarchy ownership_hierarchy() const override;
+    std::type_index element_type_index() const override;
+
+public:
+    std::list<std::string> inclusion_dependencies(
+        const formattables::inclusion_dependencies_builder_factory& f,
+        const yarn::element& e) const override;
+
+    inclusion_support_types inclusion_support_type() const override;
+
+    boost::filesystem::path inclusion_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+    boost::filesystem::path full_path(
+        const formattables::locator& l, const yarn::name& n) const override;
+
+public:
+    void register_provider(formattables::registrar& rg) const override;
+
+    dogen::formatters::file
+    format(const context& ctx, const yarn::element& e) const override;
+};
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/types/primitive_header_formatter_fwd.hpp b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/types/primitive_header_formatter_fwd.hpp
new file mode 100644
index 0000000..edff76e
--- /dev/null
+++ b/projects/quilt.cpp/include/dogen/quilt.cpp/types/formatters/types/primitive_header_formatter_fwd.hpp
@@ -0,0 +1,38 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_QUILT_CPP_TYPES_FORMATTERS_TYPES_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+#define DOGEN_QUILT_CPP_TYPES_FORMATTERS_TYPES_PRIMITIVE_HEADER_FORMATTER_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace types {
+
+class primitive_header_formatter;
+
+} } } } }
+
+#endif
diff --git a/projects/quilt.cpp/src/types/formatters/hash/primitive_header_formatter.cpp b/projects/quilt.cpp/src/types/formatters/hash/primitive_header_formatter.cpp
new file mode 100644
index 0000000..8dacb4c
--- /dev/null
+++ b/projects/quilt.cpp/src/types/formatters/hash/primitive_header_formatter.cpp
@@ -0,0 +1,2 @@
+// dummy function to suppress ranlib warnings
+void primitive_header_formatter() { }
diff --git a/projects/quilt.cpp/src/types/formatters/io/primitive_header_formatter.cpp b/projects/quilt.cpp/src/types/formatters/io/primitive_header_formatter.cpp
new file mode 100644
index 0000000..d27a279
--- /dev/null
+++ b/projects/quilt.cpp/src/types/formatters/io/primitive_header_formatter.cpp
@@ -0,0 +1,90 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#include <typeinfo>
+#include <boost/make_shared.hpp>
+#include "dogen/quilt.cpp/types/traits.hpp"
+#include "dogen/quilt.cpp/types/formatters/traits.hpp"
+#include "dogen/quilt.cpp/types/formatters/io/traits.hpp"
+#include "dogen/quilt.cpp/types/formatters/inclusion_constants.hpp"
+#include "dogen/quilt.cpp/types/formatters/assistant.hpp"
+#include "dogen/quilt.cpp/types/formatters/io/primitive_header_formatter.hpp"
+
+namespace dogen {
+namespace quilt {
+namespace cpp {
+namespace formatters {
+namespace io {
+
+std::string primitive_header_formatter::static_formatter_name() {
+    return traits::primitive_header_formatter_name();
+}
+
+std::string primitive_header_formatter::id() const {
+    static auto r(ownership_hierarchy().formatter_name());
+    return r;
+}
+
+dynamic::ownership_hierarchy
+primitive_header_formatter::ownership_hierarchy() const {
+    static dynamic::ownership_hierarchy
+        r(formatters::traits::model_name(), traits::facet_name(),
+            primitive_header_formatter::static_formatter_name(),
+            formatters::traits::header_formatter_group_name());
+    return r;
+}
+
+std::type_index primitive_header_formatter::element_type_index() const {
+    static auto r(std::type_index(typeid(yarn::object)));
+    return r;
+}
+
+std::list<std::string> primitive_header_formatter::inclusion_dependencies(
+    const formattables::inclusion_dependencies_builder_factory& /*f*/,
+    const yarn::element& /*e*/) const {
+    static const std::list<std::string> r;
+    return r;
+}
+
+inclusion_support_types
+primitive_header_formatter::inclusion_support_type() const {
+    return inclusion_support_types::canonical_support;
+}
+
+boost::filesystem::path primitive_header_formatter::inclusion_path(
+    const formattables::locator& l, const yarn::name& n) const {
+    return l.make_inclusion_path_for_cpp_header(n, static_formatter_name());
+}
+
+boost::filesystem::path primitive_header_formatter::full_path(
+    const formattables::locator& l, const yarn::name& n) const {
+    return l.make_full_path_for_cpp_header(n, static_formatter_name());
+}
+
+void primitive_header_formatter::
+register_provider(formattables::registrar& /*rg*/) const {}
+
+dogen::formatters::file primitive_header_formatter::
+format(const context& /*ctx*/, const yarn::element& /*e*/) const {
+    dogen::formatters::file r;
+    return r;
+}
+
+} } } } }
diff --git a/projects/quilt.cpp/src/types/formatters/io/traits.cpp b/projects/quilt.cpp/src/types/formatters/io/traits.cpp
index 63b87fa..25c6233 100644
--- a/projects/quilt.cpp/src/types/formatters/io/traits.cpp
+++ b/projects/quilt.cpp/src/types/formatters/io/traits.cpp
@@ -54,4 +54,9 @@ std::string traits::enum_implementation_formatter_name() {
     return r;
 }
 
+std::string traits::enum_implementation_formatter_name() {
+    static std::string r(traits::facet_name() + ".enum_implementation_formatter");
+    return r;
+}
+
 } } } } }
diff --git a/projects/quilt.cpp/src/types/formatters/odb/primitive_header_formatter.cpp b/projects/quilt.cpp/src/types/formatters/odb/primitive_header_formatter.cpp
new file mode 100644
index 0000000..8dacb4c
--- /dev/null
+++ b/projects/quilt.cpp/src/types/formatters/odb/primitive_header_formatter.cpp
@@ -0,0 +1,2 @@
+// dummy function to suppress ranlib warnings
+void primitive_header_formatter() { }
diff --git a/projects/quilt.cpp/src/types/formatters/serialization/primitive_header_formatter.cpp b/projects/quilt.cpp/src/types/formatters/serialization/primitive_header_formatter.cpp
new file mode 100644
index 0000000..8dacb4c
--- /dev/null
+++ b/projects/quilt.cpp/src/types/formatters/serialization/primitive_header_formatter.cpp
@@ -0,0 +1,2 @@
+// dummy function to suppress ranlib warnings
+void primitive_header_formatter() { }
diff --git a/projects/quilt.cpp/src/types/formatters/test_data/primitive_header_formatter.cpp b/projects/quilt.cpp/src/types/formatters/test_data/primitive_header_formatter.cpp
new file mode 100644
index 0000000..8dacb4c
--- /dev/null
+++ b/projects/quilt.cpp/src/types/formatters/test_data/primitive_header_formatter.cpp
@@ -0,0 +1,2 @@
+// dummy function to suppress ranlib warnings
+void primitive_header_formatter() { }
diff --git a/projects/quilt.cpp/src/types/formatters/types/primitive_header_formatter.cpp b/projects/quilt.cpp/src/types/formatters/types/primitive_header_formatter.cpp
new file mode 100644
index 0000000..8dacb4c
--- /dev/null
+++ b/projects/quilt.cpp/src/types/formatters/types/primitive_header_formatter.cpp
@@ -0,0 +1,2 @@
+// dummy function to suppress ranlib warnings
+void primitive_header_formatter() { }
