diff --git a/diagrams/om.dia b/diagrams/om.dia
index ceca3b1..34b5024 100644
--- a/diagrams/om.dia
+++ b/diagrams/om.dia
@@ -5260,13 +5260,13 @@ These are expected to use quotes and be on the user path.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O38">
       <dia:attribute name="obj_pos">
-        <dia:point val="-14,65"/>
+        <dia:point val="-12,65"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-14.05,64.95;3.8225,68.65"/>
+        <dia:rectangle val="-12.05,64.95;5.8225,68.65"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="-14,65"/>
+        <dia:point val="-12,65"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="17.772500000000001"/>
@@ -5617,13 +5617,13 @@ These are expected to use quotes and be on the user path.#</dia:string>
         <dia:point val="-30,47"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-30.05,46.95;0.58,52.25"/>
+        <dia:rectangle val="-30.05,46.95;-8.66,52.25"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
         <dia:point val="-30,47"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="30.530000000000001"/>
+        <dia:real val="21.289999999999999"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="5.2000000000000002"/>
@@ -5748,23 +5748,6 @@ These are expected to use quotes and be on the user path.#</dia:string>
           <dia:attribute name="parameters">
             <dia:composite type="umlparameter">
               <dia:attribute name="name">
-                <dia:string>#s#</dia:string>
-              </dia:attribute>
-              <dia:attribute name="type">
-                <dia:string>#std::ostream&amp;#</dia:string>
-              </dia:attribute>
-              <dia:attribute name="value">
-                <dia:string>##</dia:string>
-              </dia:attribute>
-              <dia:attribute name="comment">
-                <dia:string>##</dia:string>
-              </dia:attribute>
-              <dia:attribute name="kind">
-                <dia:enum val="0"/>
-              </dia:attribute>
-            </dia:composite>
-            <dia:composite type="umlparameter">
-              <dia:attribute name="name">
                 <dia:string>#t#</dia:string>
               </dia:attribute>
               <dia:attribute name="type">
@@ -5833,10 +5816,10 @@ These are expected to use quotes and be on the user path.#</dia:string>
             </dia:composite>
             <dia:composite type="umlparameter">
               <dia:attribute name="name">
-                <dia:string>#indexer#</dia:string>
+                <dia:string>#pc#</dia:string>
               </dia:attribute>
               <dia:attribute name="type">
-                <dia:string>#const sml::indexer_interface&amp;#</dia:string>
+                <dia:string>#const sml::property_cache_interface&amp;#</dia:string>
               </dia:attribute>
               <dia:attribute name="value">
                 <dia:string>##</dia:string>
@@ -5858,19 +5841,19 @@ These are expected to use quotes and be on the user path.#</dia:string>
     </dia:object>
     <dia:object type="UML - Generalization" version="1" id="O43">
       <dia:attribute name="obj_pos">
-        <dia:point val="-14.735,52.2"/>
+        <dia:point val="-19.355,52.2"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-15.585,52.15;-5.06375,65.05"/>
+        <dia:rectangle val="-20.205,52.15;-3.06375,65.05"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-14.735,52.2"/>
-        <dia:point val="-14.735,58"/>
-        <dia:point val="-5.11375,58"/>
-        <dia:point val="-5.11375,65"/>
+        <dia:point val="-19.355,52.2"/>
+        <dia:point val="-19.355,58"/>
+        <dia:point val="-3.11375,58"/>
+        <dia:point val="-3.11375,65"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -6109,7 +6092,7 @@ These are expected to use quotes and be on the user path.#</dia:string>
         <dia:point val="-35.93,34.4"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-36.68,33.65;-14.535,47.8"/>
+        <dia:rectangle val="-36.68,33.65;-19.155,47.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -6117,8 +6100,8 @@ These are expected to use quotes and be on the user path.#</dia:string>
       <dia:attribute name="orth_points">
         <dia:point val="-35.93,34.4"/>
         <dia:point val="-35.93,41"/>
-        <dia:point val="-14.735,41"/>
-        <dia:point val="-14.735,47"/>
+        <dia:point val="-19.355,41"/>
+        <dia:point val="-19.355,47"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -6141,19 +6124,19 @@ These are expected to use quotes and be on the user path.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O46">
       <dia:attribute name="obj_pos">
-        <dia:point val="-91,47"/>
+        <dia:point val="-94,47"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-91.05,46.95;-73.51,52.25"/>
+        <dia:rectangle val="-94.05,46.95;-71.12,51.45"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="-91,47"/>
+        <dia:point val="-94,47"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="17.440000000000001"/>
+        <dia:real val="22.829999999999998"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
-        <dia:real val="5.2000000000000002"/>
+        <dia:real val="4.4000000000000004"/>
       </dia:attribute>
       <dia:attribute name="name">
         <dia:string>#model_formatter_interface#</dia:string>
@@ -6252,7 +6235,7 @@ These are expected to use quotes and be on the user path.#</dia:string>
             <dia:string>##</dia:string>
           </dia:attribute>
           <dia:attribute name="type">
-            <dia:string>##</dia:string>
+            <dia:string>#file#</dia:string>
           </dia:attribute>
           <dia:attribute name="visibility">
             <dia:enum val="0"/>
@@ -6275,23 +6258,6 @@ These are expected to use quotes and be on the user path.#</dia:string>
           <dia:attribute name="parameters">
             <dia:composite type="umlparameter">
               <dia:attribute name="name">
-                <dia:string>#s#</dia:string>
-              </dia:attribute>
-              <dia:attribute name="type">
-                <dia:string>#std::ostream&amp;#</dia:string>
-              </dia:attribute>
-              <dia:attribute name="value">
-                <dia:string>##</dia:string>
-              </dia:attribute>
-              <dia:attribute name="comment">
-                <dia:string>##</dia:string>
-              </dia:attribute>
-              <dia:attribute name="kind">
-                <dia:enum val="0"/>
-              </dia:attribute>
-            </dia:composite>
-            <dia:composite type="umlparameter">
-              <dia:attribute name="name">
                 <dia:string>#m#</dia:string>
               </dia:attribute>
               <dia:attribute name="type">
@@ -6407,7 +6373,7 @@ These are expected to use quotes and be on the user path.#</dia:string>
         <dia:point val="-68,34.4"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-82.33,33.65;-67.25,47.8"/>
+        <dia:rectangle val="-82.635,33.65;-67.25,47.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -6415,8 +6381,8 @@ These are expected to use quotes and be on the user path.#</dia:string>
       <dia:attribute name="orth_points">
         <dia:point val="-68,34.4"/>
         <dia:point val="-68,41"/>
-        <dia:point val="-82.28,41"/>
-        <dia:point val="-82.28,47"/>
+        <dia:point val="-82.585,41"/>
+        <dia:point val="-82.585,47"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -6478,7 +6444,7 @@ These are expected to use quotes and be on the user path.#</dia:string>
         <dia:point val="-69,65"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-73.61,52.15;-68.25,65.8"/>
+        <dia:rectangle val="-71.22,51.35;-68.25,65.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -6486,8 +6452,8 @@ These are expected to use quotes and be on the user path.#</dia:string>
       <dia:attribute name="orth_points">
         <dia:point val="-69,65"/>
         <dia:point val="-69,58"/>
-        <dia:point val="-73.56,58"/>
-        <dia:point val="-73.56,52.2"/>
+        <dia:point val="-71.17,58"/>
+        <dia:point val="-71.17,51.4"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -7003,17 +6969,17 @@ These are expected to use quotes and be on the user path.#</dia:string>
     </dia:object>
     <dia:object type="UML - Generalization" version="1" id="O53">
       <dia:attribute name="obj_pos">
-        <dia:point val="-82.28,52.2"/>
+        <dia:point val="-82.585,51.4"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-95.5687,52.15;-81.43,65.05"/>
+        <dia:rectangle val="-95.5687,51.35;-81.735,65.05"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-82.28,52.2"/>
-        <dia:point val="-82.28,58"/>
+        <dia:point val="-82.585,51.4"/>
+        <dia:point val="-82.585,58"/>
         <dia:point val="-95.5187,58"/>
         <dia:point val="-95.5187,65"/>
       </dia:attribute>
@@ -7778,5 +7744,215 @@ These are expected to use quotes and be on the user path.#</dia:string>
         <dia:connection handle="1" to="O61" connection="1"/>
       </dia:connections>
     </dia:object>
+    <dia:object type="UML - Class" version="0" id="O63">
+      <dia:attribute name="obj_pos">
+        <dia:point val="9,47"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="8.95,46.95;31.88,51.45"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="9,47"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="22.829999999999998"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="4.4000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#module_formatter_interface#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations">
+        <dia:composite type="umloperation">
+          <dia:attribute name="name">
+            <dia:string>#format#</dia:string>
+          </dia:attribute>
+          <dia:attribute name="stereotype">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="type">
+            <dia:string>#file#</dia:string>
+          </dia:attribute>
+          <dia:attribute name="visibility">
+            <dia:enum val="0"/>
+          </dia:attribute>
+          <dia:attribute name="comment">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="abstract">
+            <dia:boolean val="true"/>
+          </dia:attribute>
+          <dia:attribute name="inheritance_type">
+            <dia:enum val="0"/>
+          </dia:attribute>
+          <dia:attribute name="query">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="class_scope">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="parameters">
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#t#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const sml::module&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#l#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const licence&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#m#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const modeline&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#marker#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const std::string&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+          </dia:attribute>
+        </dia:composite>
+      </dia:attribute>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+    </dia:object>
   </dia:layer>
 </dia:diagram>
diff --git a/projects/om/include/dogen/om/types/all.hpp b/projects/om/include/dogen/om/types/all.hpp
index 4a8bf3a..847bf54 100644
--- a/projects/om/include/dogen/om/types/all.hpp
+++ b/projects/om/include/dogen/om/types/all.hpp
@@ -46,6 +46,7 @@
 #include "dogen/om/types/modeline_group.hpp"
 #include "dogen/om/types/modeline_group_hydrator.hpp"
 #include "dogen/om/types/modeline_locations.hpp"
+#include "dogen/om/types/module_formatter_interface.hpp"
 #include "dogen/om/types/type_formatter_interface.hpp"
 #include "dogen/om/types/workflow.hpp"
 #include "dogen/om/types/workflow_error.hpp"
diff --git a/projects/om/include/dogen/om/types/cpp_types_main_header_file_formatter.hpp b/projects/om/include/dogen/om/types/cpp_types_main_header_file_formatter.hpp
index 09e89d9..7aa76fa 100644
--- a/projects/om/include/dogen/om/types/cpp_types_main_header_file_formatter.hpp
+++ b/projects/om/include/dogen/om/types/cpp_types_main_header_file_formatter.hpp
@@ -33,6 +33,7 @@
 #include "dogen/sml/types/property_cache_interface.hpp"
 #include "dogen/om/types/comment_formatter.hpp"
 #include "dogen/om/types/type_formatter_interface.hpp"
+#include "dogen/om/types/module_formatter_interface.hpp"
 
 namespace dogen {
 namespace om {
@@ -41,6 +42,7 @@ namespace om {
  * @brief Formats SML types into their C++ representation.
  */
 class cpp_types_main_header_file_formatter : public type_formatter_interface,
+                                             public module_formatter_interface,
                                              private sml::type_visitor {
 public:
     cpp_types_main_header_file_formatter(
@@ -153,6 +155,9 @@ private:
     void visit(const sml::entity& e) const override;
 
 public:
+    file format(const sml::module& module, const licence& l,
+        const modeline& modeline, const std::string& marker) const override;
+
     file format(const sml::type& t, const licence& l,
         const modeline& m, const std::string& marker,
         const sml::property_cache_interface& c) const override;
diff --git a/projects/om/include/dogen/om/types/module_formatter_interface.hpp b/projects/om/include/dogen/om/types/module_formatter_interface.hpp
new file mode 100644
index 0000000..3395825
--- /dev/null
+++ b/projects/om/include/dogen/om/types/module_formatter_interface.hpp
@@ -0,0 +1,62 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012 Kitanda <info@kitanda.co.uk>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_OM_TYPES_MODULE_FORMATTER_INTERFACE_HPP
+#define DOGEN_OM_TYPES_MODULE_FORMATTER_INTERFACE_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include <iosfwd>
+#include "dogen/sml/types/model.hpp"
+#include "dogen/sml/types/module.hpp"
+#include "dogen/om/types/file.hpp"
+#include "dogen/om/types/licence.hpp"
+#include "dogen/om/types/modeline.hpp"
+
+namespace dogen {
+namespace om {
+
+/**
+ * @brief Formatter that is responsible for the formtatting of a
+ * single type.
+ */
+class module_formatter_interface {
+public:
+    module_formatter_interface() = default;
+    module_formatter_interface(const module_formatter_interface&) = default;
+    module_formatter_interface(module_formatter_interface&&) = default;
+
+public:
+    virtual ~module_formatter_interface() noexcept { }
+
+public:
+    /**
+     * @brief Format the type according to some grammar into the
+     * stream.
+     */
+    virtual file format(const sml::module& module, const licence& l,
+        const modeline& modeline, const std::string& marker) const = 0;
+};
+
+} }
+
+#endif
diff --git a/projects/om/include/dogen/om/types/module_formatter_interface_fwd.hpp b/projects/om/include/dogen/om/types/module_formatter_interface_fwd.hpp
new file mode 100644
index 0000000..8bf001a
--- /dev/null
+++ b/projects/om/include/dogen/om/types/module_formatter_interface_fwd.hpp
@@ -0,0 +1,35 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012 Kitanda <info@kitanda.co.uk>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_OM_TYPES_MODULE_FORMATTER_INTERFACE_FWD_HPP
+#define DOGEN_OM_TYPES_MODULE_FORMATTER_INTERFACE_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace om {
+
+class module_formatter_interface;
+
+} }
+
+#endif
diff --git a/projects/om/include/dogen/om/types/type_formatter_interface.hpp b/projects/om/include/dogen/om/types/type_formatter_interface.hpp
index 818778c..acd0328 100644
--- a/projects/om/include/dogen/om/types/type_formatter_interface.hpp
+++ b/projects/om/include/dogen/om/types/type_formatter_interface.hpp
@@ -32,7 +32,6 @@
 #include "dogen/om/types/file.hpp"
 #include "dogen/om/types/licence.hpp"
 #include "dogen/om/types/modeline.hpp"
-#include "dogen/om/types/file.hpp"
 
 namespace dogen {
 namespace om {
@@ -46,7 +45,9 @@ public:
     type_formatter_interface() = default;
     type_formatter_interface(const type_formatter_interface&) = default;
     type_formatter_interface(type_formatter_interface&&) = default;
-    ~type_formatter_interface() = default;
+
+public:
+    virtual ~type_formatter_interface() noexcept { }
 
 public:
     /**
diff --git a/projects/om/src/types/cpp_types_main_header_file_formatter.cpp b/projects/om/src/types/cpp_types_main_header_file_formatter.cpp
index bcc2d9a..2c000ef 100644
--- a/projects/om/src/types/cpp_types_main_header_file_formatter.cpp
+++ b/projects/om/src/types/cpp_types_main_header_file_formatter.cpp
@@ -842,6 +842,12 @@ visit(const dogen::sml::entity& e) const {
 }
 
 file cpp_types_main_header_file_formatter::
+format(const sml::module& module, const licence& l, const modeline& modeline,
+    const std::string& marker) const {
+    
+}
+
+file cpp_types_main_header_file_formatter::
 format(const sml::type& t, const licence& l, const modeline& m,
     const std::string& marker, const sml::property_cache_interface& c) const {
 
diff --git a/projects/om/src/types/module_formatter_interface.cpp b/projects/om/src/types/module_formatter_interface.cpp
new file mode 100644
index 0000000..9d63e3e
--- /dev/null
+++ b/projects/om/src/types/module_formatter_interface.cpp
@@ -0,0 +1,2 @@
+// dummy function to suppress ranlib warnings
+void module_formatter_interface() { }
