diff --git a/diagrams/sml.dia b/diagrams/sml.dia
index 8838682..75ed5aa 100644
--- a/diagrams/sml.dia
+++ b/diagrams/sml.dia
@@ -74,7 +74,7 @@
         <dia:point val="11,60"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="32.839999999999996"/>
+        <dia:real val="32.840000000000003"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="13.200000000000001"/>
@@ -725,7 +725,7 @@ This is only needed because we don't have a type base class.#</dia:string>
         <dia:point val="-25,84"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="32.839999999999996"/>
+        <dia:real val="32.840000000000003"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="5.2000000000000011"/>
@@ -1297,7 +1297,7 @@ It must be castable to instance of the type defined in the enumeration.#</dia:st
         <dia:point val="-38,29"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="32.839999999999996"/>
+        <dia:real val="32.840000000000003"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="19.600000000000005"/>
@@ -1903,7 +1903,7 @@ or other comparable model types.#</dia:string>
         <dia:point val="-80,60"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="32.839999999999996"/>
+        <dia:real val="32.840000000000003"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="6.8000000000000007"/>
@@ -3898,7 +3898,7 @@ categories for them.#</dia:string>
         <dia:point val="-19,10"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-19.05,9.95;-9.98,17.65"/>
+        <dia:rectangle val="-19.05,9.95;-9.98,16.85"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
         <dia:point val="-19,10"/>
@@ -3907,7 +3907,7 @@ categories for them.#</dia:string>
         <dia:real val="8.9700000000000006"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
-        <dia:real val="7.6000000000000005"/>
+        <dia:real val="6.8000000000000007"/>
       </dia:attribute>
       <dia:attribute name="name">
         <dia:string>#pod_types#</dia:string>
@@ -4115,29 +4115,6 @@ which make code generation different.#</dia:string>
             <dia:boolean val="false"/>
           </dia:attribute>
         </dia:composite>
-        <dia:composite type="umlattribute">
-          <dia:attribute name="name">
-            <dia:string>#service#</dia:string>
-          </dia:attribute>
-          <dia:attribute name="type">
-            <dia:string>##</dia:string>
-          </dia:attribute>
-          <dia:attribute name="value">
-            <dia:string>##</dia:string>
-          </dia:attribute>
-          <dia:attribute name="comment">
-            <dia:string>#Pod is a service#</dia:string>
-          </dia:attribute>
-          <dia:attribute name="visibility">
-            <dia:enum val="0"/>
-          </dia:attribute>
-          <dia:attribute name="abstract">
-            <dia:boolean val="false"/>
-          </dia:attribute>
-          <dia:attribute name="class_scope">
-            <dia:boolean val="false"/>
-          </dia:attribute>
-        </dia:composite>
       </dia:attribute>
       <dia:attribute name="operations"/>
       <dia:attribute name="template">
@@ -4186,7 +4163,7 @@ which make code generation different.#</dia:string>
         <dia:point val="-5.16,29"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-14.565,17.55;-4.41,29.8"/>
+        <dia:rectangle val="-14.565,16.75;-4.41,29.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -4195,7 +4172,7 @@ which make code generation different.#</dia:string>
         <dia:point val="-5.16,29"/>
         <dia:point val="-5.16,23"/>
         <dia:point val="-14.515,23"/>
-        <dia:point val="-14.515,17.6"/>
+        <dia:point val="-14.515,16.8"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -4549,7 +4526,7 @@ which make code generation different.#</dia:string>
         <dia:point val="-32,60"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="8.3349999999999991"/>
+        <dia:real val="8.3350000000000009"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="5.2000000000000011"/>
@@ -6349,7 +6326,7 @@ and all of the supported types within those models.#</dia:string>
         <dia:real val="24.77"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
-        <dia:real val="4.0999999999999996"/>
+        <dia:real val="4.1000000000000005"/>
       </dia:attribute>
       <dia:attribute name="line_width">
         <dia:real val="0.10000000000000001"/>
@@ -6723,7 +6700,7 @@ all types in the combined model.#</dia:string>
         <dia:point val="5,11"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="31.685000000000002"/>
+        <dia:real val="31.684999999999999"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="7.5999999999999996"/>
@@ -7306,10 +7283,10 @@ all types in the combined model.#</dia:string>
         <dia:point val="-79,-11"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="27.850000000000001"/>
+        <dia:real val="27.849999999999998"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
-        <dia:real val="7.2999999999999998"/>
+        <dia:real val="7.3000000000000007"/>
       </dia:attribute>
       <dia:attribute name="line_width">
         <dia:real val="0.10000000000000001"/>
@@ -7771,7 +7748,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="-32,-6"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="7.0924999999999994"/>
+        <dia:real val="7.0925000000000002"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="3"/>
@@ -7911,7 +7888,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="-8.05,-6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-28.5038,-11.75;-7.3,-5.2"/>
+        <dia:rectangle val="-28.5037,-11.75;-7.3,-5.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -7982,7 +7959,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="-39.5525,89.2"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-43.1425,88.45;-38.8025,99.6"/>
+        <dia:rectangle val="-43.1425,88.45;-38.8025,98.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -8053,7 +8030,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="-21.58,48.6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-27.8825,47.85;-20.83,61.6"/>
+        <dia:rectangle val="-27.8825,47.85;-20.83,60.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -8124,7 +8101,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="-38,29"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-46.84,22.95;-37.25,30.6"/>
+        <dia:rectangle val="-46.84,22.25;-37.25,29.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -8195,7 +8172,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="36.685,11"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="26.215,0.75;37.435,12.6"/>
+        <dia:rectangle val="26.215,0.75;37.435,11.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -8266,7 +8243,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="36.685,11"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="35.935,-1.65;56.4975,12.6"/>
+        <dia:rectangle val="35.935,-1.65;56.4975,11.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -8337,7 +8314,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="36.685,11"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="35.935,-0.05;81.6525,12.6"/>
+        <dia:rectangle val="35.935,-0.05;81.6525,11.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -8378,7 +8355,7 @@ but with a focus only on data structures and domain driven constructs.
         <dia:point val="68,84"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="32.839999999999996"/>
+        <dia:real val="32.840000000000003"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="12.4"/>
@@ -8803,7 +8780,7 @@ Only applicable if is_parent is true.#</dia:string>
         <dia:point val="27.42,73.2"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="26.67,72.45;84.62,85.6"/>
+        <dia:rectangle val="26.67,72.45;84.62,84.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -8844,7 +8821,7 @@ Only applicable if is_parent is true.#</dia:string>
         <dia:point val="68,110"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="6.6225000000000005"/>
+        <dia:real val="6.6224999999999996"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
         <dia:real val="5.2000000000000011"/>
@@ -9054,7 +9031,7 @@ Only applicable if is_parent is true.#</dia:string>
         <dia:point val="84.42,96.4"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="71.2613,95.65;85.17,111.6"/>
+        <dia:rectangle val="71.2613,95.65;85.17,110.8"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
diff --git a/projects/dia_to_sml/src/types/transformer.cpp b/projects/dia_to_sml/src/types/transformer.cpp
index 0a0d9ad..41841c5 100644
--- a/projects/dia_to_sml/src/types/transformer.cpp
+++ b/projects/dia_to_sml/src/types/transformer.cpp
@@ -204,8 +204,6 @@ transform_pod(const object_profile& op, const processed_object& po) {
     pod.pod_type(pod_types::value);
     if (op.is_entity())
         pod.pod_type(pod_types::entity);
-    else if (op.is_service()) // FIXME: service HACK
-        pod.pod_type(pod_types::service);
 
     pod.is_fluent(op.is_fluent());
     pod.is_versioned(op.is_versioned());
@@ -217,9 +215,7 @@ transform_pod(const object_profile& op, const processed_object& po) {
         generation_types::full_generation :
         generation_types::no_generation);
 
-    // FIXME: service hack
-    // if (context_.is_target() && op.is_non_generatable())
-    if (context_.is_target() && (op.is_non_generatable() || op.is_service()))
+    if (context_.is_target() && op.is_non_generatable())
         pod.generation_type(generation_types::partial_generation);
 
     const auto pair(comments_parser_->parse(po.comment()));
@@ -560,9 +556,8 @@ transform(const processed_object& o, const object_profile& op) {
         transform_enumeration(o);
     else if (op.is_exception())
         transform_exception(o);
-    // FIXME: service hack
-    // else if (op.is_service())
-    //     transform_service(o);
+    else if (op.is_service())
+        transform_service(o);
     else
         transform_pod(op, o);
 
diff --git a/projects/sml/include/dogen/sml/types/pod_types.hpp b/projects/sml/include/dogen/sml/types/pod_types.hpp
index 1633271..e4d0e5d 100644
--- a/projects/sml/include/dogen/sml/types/pod_types.hpp
+++ b/projects/sml/include/dogen/sml/types/pod_types.hpp
@@ -40,8 +40,7 @@ enum class pod_types : unsigned int {
     associative_container = 2, ///< Pod is an associative container
     smart_pointer = 3, ///< Pod is a smart pointer
     entity = 4, ///< Pod is a domain entity
-    value = 5, ///< Pod is a domain value
-    service = 6 ///< Pod is a service
+    value = 5 ///< Pod is a domain value
 };
 
 } }
diff --git a/projects/sml/src/io/pod_types_io.cpp b/projects/sml/src/io/pod_types_io.cpp
index b4b0425..d08ecb7 100644
--- a/projects/sml/src/io/pod_types_io.cpp
+++ b/projects/sml/src/io/pod_types_io.cpp
@@ -48,9 +48,6 @@ std::ostream& operator<<(std::ostream& s, const pod_types& v) {
     case pod_types::value:
         attr = "\"value\"";
         break;
-    case pod_types::service:
-        attr = "\"service\"";
-        break;
     default:
         throw std::invalid_argument("Invalid value for pod_types");
     }
diff --git a/projects/sml/src/test_data/pod_types_td.cpp b/projects/sml/src/test_data/pod_types_td.cpp
index e63efd6..25d6c85 100644
--- a/projects/sml/src/test_data/pod_types_td.cpp
+++ b/projects/sml/src/test_data/pod_types_td.cpp
@@ -26,7 +26,7 @@ namespace sml {
 pod_types_generator::pod_types_generator() : position_(0) { }
 void pod_types_generator::
 populate(const unsigned int position, result_type& v) {
-    v = static_cast<pod_types>(position % 7);
+    v = static_cast<pod_types>(position % 6);
 }
 
 pod_types_generator::result_type
