diff --git a/projects/generator/src/backends/cpp/formatters/cpp_class_declaration.cpp b/projects/generator/src/backends/cpp/formatters/cpp_class_declaration.cpp
index 33dce4d..27c345b 100644
--- a/projects/generator/src/backends/cpp/formatters/cpp_class_declaration.cpp
+++ b/projects/generator/src/backends/cpp/formatters/cpp_class_declaration.cpp
@@ -32,7 +32,25 @@ cpp_class_declaration::cpp_class_declaration(std::ostream& stream)
     : stream_(stream), utility_(stream_, indenter_) { }
 
 void cpp_class_declaration::open_class(class_view_model vm) {
-    stream_ << indenter_ << "class " << vm.name() << " {" << std::endl;
+    stream_ << indenter_ << "class " << vm.name();
+
+    const auto parents(vm.parents());
+    if (!parents.empty()) {
+        stream_ << " :";
+
+        if (parents.size() == 1) {
+            stream_ << " public " << parents.front().name();
+        } else {
+            bool is_first(true);
+            cpp_positive_indenter_scope s(indenter_);
+            for (const auto p : vm.parents()) {
+                stream_ << (is_first ? "" : ",") << std::endl << indenter_;
+                stream_ << " public " << p.name();
+                is_first = false;
+            }
+        }
+    }
+    stream_ << " {" << std::endl;
 }
 
 void cpp_class_declaration::close_class() {
