diff --git a/diagrams/formatters.dia b/diagrams/formatters.dia
index b50f65c..02f5709 100644
--- a/diagrams/formatters.dia
+++ b/diagrams/formatters.dia
@@ -65,13 +65,13 @@
   <dia:layer name="Background" visible="true" active="true">
     <dia:object type="UML - Class" version="0" id="O0">
       <dia:attribute name="obj_pos">
-        <dia:point val="-51,-23"/>
+        <dia:point val="-51,-22"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-51.05,-23.05;-37.36,-17.75"/>
+        <dia:rectangle val="-51.05,-22.05;-37.36,-16.75"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="-51,-23"/>
+        <dia:point val="-51,-22"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="13.59"/>
@@ -277,13 +277,13 @@ separator is @e ;.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O1">
       <dia:attribute name="obj_pos">
-        <dia:point val="-12,-23"/>
+        <dia:point val="-12,-22"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-12.05,-23.05;4.72,-19.35"/>
+        <dia:rectangle val="-12.05,-22.05;4.72,-18.35"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="-12,-23"/>
+        <dia:point val="-12,-22"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="16.669999999999998"/>
@@ -641,13 +641,13 @@ separator is @e ;.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O3">
       <dia:attribute name="obj_pos">
-        <dia:point val="-30,-38"/>
+        <dia:point val="-31,-38"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-30.05,-38.05;-11.74,-33.55"/>
+        <dia:rectangle val="-31.05,-38.05;-12.74,-33.55"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="-30,-38"/>
+        <dia:point val="-31,-38"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="18.210000000000001"/>
@@ -1112,13 +1112,13 @@ separator is @e ;.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O6">
       <dia:attribute name="obj_pos">
-        <dia:point val="-31,-23"/>
+        <dia:point val="-31,-22"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-31.05,-23.05;-23.135,-19.35"/>
+        <dia:rectangle val="-31.05,-22.05;-23.135,-18.35"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="-31,-23"/>
+        <dia:point val="-31,-22"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="7.8150000000000004"/>
@@ -1305,19 +1305,19 @@ separator is @e ;.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="-37.41,-17.8"/>
+        <dia:point val="-37.41,-16.8"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-38.16,-19.45;-26.8925,-10.95"/>
+        <dia:rectangle val="-38.16,-18.45;-26.8925,-10.95"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-37.41,-17.8"/>
+        <dia:point val="-37.41,-16.8"/>
         <dia:point val="-37.41,-11"/>
         <dia:point val="-27.0925,-11"/>
-        <dia:point val="-27.0925,-19.4"/>
+        <dia:point val="-27.0925,-18.4"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -1539,16 +1539,16 @@ For example, one may choose to have a set of modelines for @e emacs , or for
         <dia:point val="-45.74,-34.4"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-46.49,-35.15;-44.005,-22.2"/>
+        <dia:rectangle val="-46.49,-35.15;-44.005,-21.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
         <dia:point val="-45.74,-34.4"/>
-        <dia:point val="-45.74,-28"/>
-        <dia:point val="-44.205,-28"/>
-        <dia:point val="-44.205,-23"/>
+        <dia:point val="-45.74,-27"/>
+        <dia:point val="-44.205,-27"/>
+        <dia:point val="-44.205,-22"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -2219,16 +2219,16 @@ For example, one may choose to have a set of modelines for @e emacs , or for
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="-51,-17.8"/>
+        <dia:point val="-51,-16.8"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-69.4737,-18.55;-50.25,-4.2"/>
+        <dia:rectangle val="-69.4737,-17.55;-50.25,-4.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-51,-17.8"/>
+        <dia:point val="-51,-16.8"/>
         <dia:point val="-51,-11"/>
         <dia:point val="-69.4237,-11"/>
         <dia:point val="-69.4237,-5"/>
@@ -2290,18 +2290,18 @@ For example, one may choose to have a set of modelines for @e emacs , or for
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="-51,-23"/>
+        <dia:point val="-51,-22"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-63.9125,-32.05;-50.25,-22.2"/>
+        <dia:rectangle val="-63.9125,-32.05;-50.25,-21.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-51,-23"/>
-        <dia:point val="-51,-28"/>
-        <dia:point val="-63.8625,-28"/>
+        <dia:point val="-51,-22"/>
+        <dia:point val="-51,-27"/>
+        <dia:point val="-63.8625,-27"/>
         <dia:point val="-63.8625,-32"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
@@ -2879,16 +2879,16 @@ only be written to the filesystem if it does not exist.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="-44.205,-17.8"/>
+        <dia:point val="-44.205,-16.8"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-46.2875,-18.55;-43.455,-4.2"/>
+        <dia:rectangle val="-46.2875,-17.55;-43.455,-4.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-44.205,-17.8"/>
+        <dia:point val="-44.205,-16.8"/>
         <dia:point val="-44.205,-11"/>
         <dia:point val="-46.2375,-11"/>
         <dia:point val="-46.2375,-5"/>
@@ -3398,13 +3398,13 @@ only be written to the filesystem if it does not exist.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O23">
       <dia:attribute name="obj_pos">
-        <dia:point val="16,-5"/>
+        <dia:point val="18,-5"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="15.95,-5.05;23.8075,1.05"/>
+        <dia:rectangle val="17.95,-5.05;25.8075,1.05"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="16,-5"/>
+        <dia:point val="18,-5"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="7.7575000000000003"/>
@@ -3640,7 +3640,7 @@ only be written to the filesystem if it does not exist.#</dia:string>
         <dia:point val="-1.43,-5"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-2.18,-11.75;16.2,-4.2"/>
+        <dia:rectangle val="-2.18,-11.75;18.2,-4.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -3648,8 +3648,8 @@ only be written to the filesystem if it does not exist.#</dia:string>
       <dia:attribute name="orth_points">
         <dia:point val="-1.43,-5"/>
         <dia:point val="-1.43,-11"/>
-        <dia:point val="16,-11"/>
-        <dia:point val="16,-5"/>
+        <dia:point val="18,-11"/>
+        <dia:point val="18,-5"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -3936,19 +3936,19 @@ only be written to the filesystem if it does not exist.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="-0.82,-33.6"/>
+        <dia:point val="0.18,-33.6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-3.715,-34.35;-0.07,-22.2"/>
+        <dia:rectangle val="-3.715,-34.35;0.93,-21.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-0.82,-33.6"/>
-        <dia:point val="-0.82,-28"/>
-        <dia:point val="-3.665,-28"/>
-        <dia:point val="-3.665,-23"/>
+        <dia:point val="0.18,-33.6"/>
+        <dia:point val="0.18,-27"/>
+        <dia:point val="-3.665,-27"/>
+        <dia:point val="-3.665,-22"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -4007,17 +4007,17 @@ only be written to the filesystem if it does not exist.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="47.955,-47.6"/>
+        <dia:point val="35,-47.6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="23.69,-48.35;48.705,-37.2"/>
+        <dia:rectangle val="23.69,-48.35;35.75,-37.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="47.955,-47.6"/>
-        <dia:point val="47.955,-43"/>
+        <dia:point val="35,-47.6"/>
+        <dia:point val="35,-43"/>
         <dia:point val="23.74,-43"/>
         <dia:point val="23.74,-38"/>
       </dia:attribute>
@@ -4036,7 +4036,7 @@ only be written to the filesystem if it does not exist.#</dia:string>
         <dia:color val="#000000"/>
       </dia:attribute>
       <dia:connections>
-        <dia:connection handle="0" to="O38" connection="6"/>
+        <dia:connection handle="0" to="O38" connection="5"/>
         <dia:connection handle="1" to="O26" connection="2"/>
       </dia:connections>
     </dia:object>
@@ -4078,19 +4078,19 @@ only be written to the filesystem if it does not exist.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="-8,-33.6"/>
+        <dia:point val="-7,-33.6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-37.46,-34.35;-7.25,-22.2"/>
+        <dia:rectangle val="-37.46,-34.35;-6.25,-21.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="-8,-33.6"/>
-        <dia:point val="-8,-28"/>
-        <dia:point val="-37.41,-28"/>
-        <dia:point val="-37.41,-23"/>
+        <dia:point val="-7,-33.6"/>
+        <dia:point val="-7,-27"/>
+        <dia:point val="-37.41,-27"/>
+        <dia:point val="-37.41,-22"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -4513,13 +4513,13 @@ then make use of it.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O36">
       <dia:attribute name="obj_pos">
-        <dia:point val="-8,-38"/>
+        <dia:point val="-7,-38"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-8.05,-38.05;6.41,-33.55"/>
+        <dia:rectangle val="-7.05,-38.05;7.41,-33.55"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="-8,-38"/>
+        <dia:point val="-7,-38"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="14.359999999999999"/>
@@ -4732,7 +4732,7 @@ then make use of it.
         <dia:point val="14,-38"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="-0.87,-43.75;14.75,-37.2"/>
+        <dia:rectangle val="0.13,-43.75;14.75,-37.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -4740,8 +4740,8 @@ then make use of it.
       <dia:attribute name="orth_points">
         <dia:point val="14,-38"/>
         <dia:point val="14,-43"/>
-        <dia:point val="-0.82,-43"/>
-        <dia:point val="-0.82,-38.0504"/>
+        <dia:point val="0.18,-43"/>
+        <dia:point val="0.18,-38.0504"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -4957,13 +4957,13 @@ then make use of it.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O39">
       <dia:attribute name="obj_pos">
-        <dia:point val="73,-70"/>
+        <dia:point val="67,-64"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="72.95,-70.05;79.91,-66.95"/>
+        <dia:rectangle val="66.95,-64.05;73.91,-60.95"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="73,-70"/>
+        <dia:point val="67,-64"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="6.8600000000000003"/>
@@ -5067,7 +5067,7 @@ then make use of it.
     </dia:object>
     <dia:object type="UML - Association" version="2" id="O40">
       <dia:attribute name="name">
-        <dia:string>#on error throws &gt;#</dia:string>
+        <dia:string>##</dia:string>
       </dia:attribute>
       <dia:attribute name="direction">
         <dia:enum val="2"/>
@@ -5103,18 +5103,18 @@ then make use of it.
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="73,-67"/>
+        <dia:point val="67,-61"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="60.16,-67.05;73.2,-51.2"/>
+        <dia:rectangle val="60.16,-61.05;67.2,-51.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="73,-67"/>
-        <dia:point val="73,-61"/>
-        <dia:point val="60.91,-61"/>
+        <dia:point val="67,-61"/>
+        <dia:point val="67,-57"/>
+        <dia:point val="60.91,-57"/>
         <dia:point val="60.91,-52"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
@@ -5138,13 +5138,13 @@ then make use of it.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O41">
       <dia:attribute name="obj_pos">
-        <dia:point val="105,-70"/>
+        <dia:point val="90,-70"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="104.95,-70.05;119.025,-61.55"/>
+        <dia:rectangle val="89.95,-70.05;104.025,-61.55"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="105,-70"/>
+        <dia:point val="90,-70"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="13.975"/>
@@ -5495,19 +5495,19 @@ then make use of it.
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="116.14,-52"/>
+        <dia:point val="92.57,-52"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="111.937,-61.65;116.89,-51.2"/>
+        <dia:rectangle val="91.82,-61.65;97.1875,-51.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="116.14,-52"/>
-        <dia:point val="116.14,-56"/>
-        <dia:point val="111.987,-56"/>
-        <dia:point val="111.987,-61.6"/>
+        <dia:point val="92.57,-52"/>
+        <dia:point val="92.57,-57"/>
+        <dia:point val="96.9875,-57"/>
+        <dia:point val="96.9875,-61.6"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -5524,7 +5524,7 @@ then make use of it.
         <dia:color val="#000000"/>
       </dia:attribute>
       <dia:connections>
-        <dia:connection handle="0" to="O82" connection="2"/>
+        <dia:connection handle="0" to="O82" connection="1"/>
         <dia:connection handle="1" to="O41" connection="6"/>
       </dia:connections>
     </dia:object>
@@ -5971,7 +5971,7 @@ then make use of it.
     </dia:object>
     <dia:object type="UML - Association" version="2" id="O48">
       <dia:attribute name="name">
-        <dia:string>#queries &gt;#</dia:string>
+        <dia:string>#&lt; queries#</dia:string>
       </dia:attribute>
       <dia:attribute name="direction">
         <dia:enum val="1"/>
@@ -6017,8 +6017,8 @@ then make use of it.
       </dia:attribute>
       <dia:attribute name="orth_points">
         <dia:point val="35,-52"/>
-        <dia:point val="35,-61"/>
-        <dia:point val="27.14,-61"/>
+        <dia:point val="35,-57"/>
+        <dia:point val="27.14,-57"/>
         <dia:point val="27.14,-66.4"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
@@ -6809,22 +6809,22 @@ then make use of it.
     </dia:object>
     <dia:object type="UML - LargePackage" version="0" id="O54">
       <dia:attribute name="obj_pos">
-        <dia:point val="54,-31"/>
+        <dia:point val="56,-29"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="53.95,-32.05;148.05,22.05"/>
+        <dia:rectangle val="55.95,-30.05;148.05,22.05"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="54,-31"/>
+        <dia:point val="56,-29"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
-        <dia:real val="94"/>
+        <dia:real val="92"/>
       </dia:attribute>
       <dia:attribute name="elem_height">
-        <dia:real val="53"/>
+        <dia:real val="51"/>
       </dia:attribute>
       <dia:attribute name="line_width">
         <dia:real val="0.10000000149011612"/>
@@ -6847,13 +6847,13 @@ then make use of it.
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O55">
       <dia:attribute name="obj_pos">
-        <dia:point val="74,-23"/>
+        <dia:point val="74,-22"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="73.95,-23.05;99.19,-16.15"/>
+        <dia:rectangle val="73.95,-22.05;99.19,-15.15"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="74,-23"/>
+        <dia:point val="74,-22"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="25.140000000000001"/>
@@ -7612,18 +7612,18 @@ then make use of it.
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="74,-16.2"/>
+        <dia:point val="74,-15.2"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="67.44,-16.95;74.75,-5.2"/>
+        <dia:rectangle val="67.44,-15.95;74.75,-5.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="74,-16.2"/>
-        <dia:point val="74,-10.75"/>
-        <dia:point val="67.49,-10.75"/>
+        <dia:point val="74,-15.2"/>
+        <dia:point val="74,-11"/>
+        <dia:point val="67.49,-11"/>
         <dia:point val="67.49,-6"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
@@ -7632,7 +7632,7 @@ then make use of it.
         <dia:enum val="1"/>
       </dia:attribute>
       <dia:attribute name="orth_autoroute">
-        <dia:boolean val="true"/>
+        <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="text_colour">
         <dia:color val="#000000"/>
@@ -7684,18 +7684,18 @@ then make use of it.
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="99.14,-16.2"/>
+        <dia:point val="99.14,-15.2"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="98.39,-16.95;104.578,-5.2"/>
+        <dia:rectangle val="98.39,-15.95;104.578,-5.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="99.14,-16.2"/>
-        <dia:point val="99.14,-10.75"/>
-        <dia:point val="104.377,-10.75"/>
+        <dia:point val="99.14,-15.2"/>
+        <dia:point val="99.14,-11"/>
+        <dia:point val="104.377,-11"/>
         <dia:point val="104.377,-6"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
@@ -7704,7 +7704,7 @@ then make use of it.
         <dia:enum val="1"/>
       </dia:attribute>
       <dia:attribute name="orth_autoroute">
-        <dia:boolean val="true"/>
+        <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="text_colour">
         <dia:color val="#000000"/>
@@ -9645,7 +9645,7 @@ then make use of it.
         <dia:point val="129.265,-23"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="99.09,-27.75;130.015,-22.2"/>
+        <dia:rectangle val="99.09,-27.75;130.015,-21.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
@@ -9654,7 +9654,7 @@ then make use of it.
         <dia:point val="129.265,-23"/>
         <dia:point val="129.265,-27"/>
         <dia:point val="99.14,-27"/>
-        <dia:point val="99.14,-23"/>
+        <dia:point val="99.14,-22"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -10507,13 +10507,13 @@ than one element.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O77">
       <dia:attribute name="obj_pos">
-        <dia:point val="13,-23"/>
+        <dia:point val="13,-22"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="12.95,-23.05;33.57,-16.95"/>
+        <dia:rectangle val="12.95,-22.05;33.57,-15.95"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="13,-23"/>
+        <dia:point val="13,-22"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="20.52"/>
@@ -10818,19 +10818,19 @@ than one element.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="23.26,-17"/>
+        <dia:point val="23.26,-16"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="22.51,-17.75;24.01,-4.2"/>
+        <dia:rectangle val="22.51,-16.75;25.9575,-4.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="23.26,-17"/>
+        <dia:point val="23.26,-16"/>
         <dia:point val="23.26,-11"/>
-        <dia:point val="23.7575,-11"/>
-        <dia:point val="23.7575,-5"/>
+        <dia:point val="25.7575,-11"/>
+        <dia:point val="25.7575,-5"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -10889,19 +10889,19 @@ than one element.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="6.36,-33.6"/>
+        <dia:point val="7.36,-33.6"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="6.31,-33.65;13.75,-22.2"/>
+        <dia:rectangle val="7.31,-33.65;13.75,-21.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="6.36,-33.6"/>
-        <dia:point val="6.36,-27.95"/>
-        <dia:point val="13,-27.95"/>
-        <dia:point val="13,-23"/>
+        <dia:point val="7.36,-33.6"/>
+        <dia:point val="7.36,-27"/>
+        <dia:point val="13,-27"/>
+        <dia:point val="13,-22"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -10960,19 +10960,19 @@ than one element.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="74,-23"/>
+        <dia:point val="74,-22"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="33.47,-28.75;74.75,-22.2"/>
+        <dia:rectangle val="33.47,-25.75;74.75,-21.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="74,-23"/>
-        <dia:point val="74,-28"/>
-        <dia:point val="33.52,-28"/>
-        <dia:point val="33.52,-23"/>
+        <dia:point val="74,-22"/>
+        <dia:point val="74,-25"/>
+        <dia:point val="33.52,-25"/>
+        <dia:point val="33.52,-22"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -10995,13 +10995,13 @@ than one element.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O81">
       <dia:attribute name="obj_pos">
-        <dia:point val="67,-52"/>
+        <dia:point val="114,-52"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="66.95,-52.05;84.49,-44.35"/>
+        <dia:rectangle val="113.95,-52.05;131.49,-44.35"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="67,-52"/>
+        <dia:point val="114,-52"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="17.440000000000001"/>
@@ -11267,13 +11267,13 @@ than one element.#</dia:string>
     </dia:object>
     <dia:object type="UML - Class" version="0" id="O82">
       <dia:attribute name="obj_pos">
-        <dia:point val="91,-52"/>
+        <dia:point val="80,-52"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="90.95,-52.05;116.19,-48.35"/>
+        <dia:rectangle val="79.95,-52.05;105.19,-48.35"/>
       </dia:attribute>
       <dia:attribute name="elem_corner">
-        <dia:point val="91,-52"/>
+        <dia:point val="80,-52"/>
       </dia:attribute>
       <dia:attribute name="elem_width">
         <dia:real val="25.140000000000001"/>
@@ -11462,19 +11462,19 @@ than one element.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="103.57,-48.4"/>
+        <dia:point val="105.14,-52"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="75.67,-49.15;104.32,-38.95"/>
+        <dia:rectangle val="104.39,-57.75;122.92,-51.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="103.57,-48.4"/>
-        <dia:point val="103.57,-39"/>
-        <dia:point val="75.72,-39"/>
-        <dia:point val="75.72,-44.4"/>
+        <dia:point val="105.14,-52"/>
+        <dia:point val="105.14,-57"/>
+        <dia:point val="122.72,-57"/>
+        <dia:point val="122.72,-52"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -11491,8 +11491,8 @@ than one element.#</dia:string>
         <dia:color val="#000000"/>
       </dia:attribute>
       <dia:connections>
-        <dia:connection handle="0" to="O82" connection="6"/>
-        <dia:connection handle="1" to="O81" connection="6"/>
+        <dia:connection handle="0" to="O82" connection="2"/>
+        <dia:connection handle="1" to="O81" connection="1"/>
       </dia:connections>
     </dia:object>
     <dia:object type="UML - Association" version="2" id="O84">
@@ -11533,19 +11533,19 @@ than one element.#</dia:string>
         <dia:boolean val="false"/>
       </dia:attribute>
       <dia:attribute name="obj_pos">
-        <dia:point val="91,-52"/>
+        <dia:point val="80,-52"/>
       </dia:attribute>
       <dia:attribute name="obj_bb">
-        <dia:rectangle val="79.81,-67.05;91.75,-51.2"/>
+        <dia:rectangle val="73.81,-61.05;80.75,-51.2"/>
       </dia:attribute>
       <dia:attribute name="meta">
         <dia:composite type="dict"/>
       </dia:attribute>
       <dia:attribute name="orth_points">
-        <dia:point val="91,-52"/>
-        <dia:point val="91,-61"/>
-        <dia:point val="79.86,-61"/>
-        <dia:point val="79.86,-67"/>
+        <dia:point val="80,-52"/>
+        <dia:point val="80,-57"/>
+        <dia:point val="73.86,-57"/>
+        <dia:point val="73.86,-61"/>
       </dia:attribute>
       <dia:attribute name="orth_orient">
         <dia:enum val="1"/>
@@ -11566,5 +11566,507 @@ than one element.#</dia:string>
         <dia:connection handle="1" to="O39" connection="7"/>
       </dia:connections>
     </dia:object>
+    <dia:object type="UML - Class" version="0" id="O85">
+      <dia:attribute name="obj_pos">
+        <dia:point val="41,-38"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="40.95,-38.05;63.88,-31.15"/>
+      </dia:attribute>
+      <dia:attribute name="elem_corner">
+        <dia:point val="41,-38"/>
+      </dia:attribute>
+      <dia:attribute name="elem_width">
+        <dia:real val="22.830000000000002"/>
+      </dia:attribute>
+      <dia:attribute name="elem_height">
+        <dia:real val="6.8000000000000007"/>
+      </dia:attribute>
+      <dia:attribute name="name">
+        <dia:string>#file_properties_workflow#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="stereotype">
+        <dia:string>#service#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="comment">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="abstract">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_attributes">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="suppress_operations">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="visible_attributes">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="visible_comments">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_operations">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="wrap_after_char">
+        <dia:int val="40"/>
+      </dia:attribute>
+      <dia:attribute name="comment_line_length">
+        <dia:int val="17"/>
+      </dia:attribute>
+      <dia:attribute name="comment_tagging">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="line_width">
+        <dia:real val="0.10000000000000001"/>
+      </dia:attribute>
+      <dia:attribute name="line_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="fill_color">
+        <dia:color val="#ffffff"/>
+      </dia:attribute>
+      <dia:attribute name="text_color">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font">
+        <dia:font family="monospace" style="0" name="Courier"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font">
+        <dia:font family="monospace" style="88" name="Courier-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font">
+        <dia:font family="monospace" style="8" name="Courier-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font">
+        <dia:font family="sans" style="80" name="Helvetica-Bold"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font">
+        <dia:font family="sans" style="88" name="Helvetica-BoldOblique"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font">
+        <dia:font family="sans" style="8" name="Helvetica-Oblique"/>
+      </dia:attribute>
+      <dia:attribute name="normal_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="polymorphic_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_font_height">
+        <dia:real val="0.80000000000000004"/>
+      </dia:attribute>
+      <dia:attribute name="classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="abstract_classname_font_height">
+        <dia:real val="1"/>
+      </dia:attribute>
+      <dia:attribute name="comment_font_height">
+        <dia:real val="0.69999999999999996"/>
+      </dia:attribute>
+      <dia:attribute name="attributes"/>
+      <dia:attribute name="operations">
+        <dia:composite type="umloperation">
+          <dia:attribute name="name">
+            <dia:string>#file_properties_workflow#</dia:string>
+          </dia:attribute>
+          <dia:attribute name="stereotype">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="type">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="visibility">
+            <dia:enum val="0"/>
+          </dia:attribute>
+          <dia:attribute name="comment">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="abstract">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="inheritance_type">
+            <dia:enum val="2"/>
+          </dia:attribute>
+          <dia:attribute name="query">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="class_scope">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="parameters">
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#rp#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const repository&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+          </dia:attribute>
+        </dia:composite>
+        <dia:composite type="umloperation">
+          <dia:attribute name="name">
+            <dia:string>#file_properties_workflow#</dia:string>
+          </dia:attribute>
+          <dia:attribute name="stereotype">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="type">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="visibility">
+            <dia:enum val="0"/>
+          </dia:attribute>
+          <dia:attribute name="comment">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="abstract">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="inheritance_type">
+            <dia:enum val="2"/>
+          </dia:attribute>
+          <dia:attribute name="query">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="class_scope">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="parameters">
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#rp#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const repository&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#fallback#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const dynamic::object&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+          </dia:attribute>
+        </dia:composite>
+        <dia:composite type="umloperation">
+          <dia:attribute name="name">
+            <dia:string>#execute#</dia:string>
+          </dia:attribute>
+          <dia:attribute name="stereotype">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="type">
+            <dia:string>#file_properties#</dia:string>
+          </dia:attribute>
+          <dia:attribute name="visibility">
+            <dia:enum val="0"/>
+          </dia:attribute>
+          <dia:attribute name="comment">
+            <dia:string>##</dia:string>
+          </dia:attribute>
+          <dia:attribute name="abstract">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="inheritance_type">
+            <dia:enum val="2"/>
+          </dia:attribute>
+          <dia:attribute name="query">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="class_scope">
+            <dia:boolean val="false"/>
+          </dia:attribute>
+          <dia:attribute name="parameters">
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#modeline_name#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const std::string&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+            <dia:composite type="umlparameter">
+              <dia:attribute name="name">
+                <dia:string>#o#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="type">
+                <dia:string>#const dynamic::object&amp;#</dia:string>
+              </dia:attribute>
+              <dia:attribute name="value">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="comment">
+                <dia:string>##</dia:string>
+              </dia:attribute>
+              <dia:attribute name="kind">
+                <dia:enum val="0"/>
+              </dia:attribute>
+            </dia:composite>
+          </dia:attribute>
+        </dia:composite>
+      </dia:attribute>
+      <dia:attribute name="template">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="templates"/>
+    </dia:object>
+    <dia:object type="UML - Association" version="2" id="O86">
+      <dia:attribute name="name">
+        <dia:string>#&lt; produces#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="direction">
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="show_direction">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="assoc_type">
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="role_a">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="multipicity_a">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="visibility_a">
+        <dia:enum val="3"/>
+      </dia:attribute>
+      <dia:attribute name="show_arrow_a">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="role_b">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="multipicity_b">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="visibility_b">
+        <dia:enum val="3"/>
+      </dia:attribute>
+      <dia:attribute name="show_arrow_b">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="obj_pos">
+        <dia:point val="52.415,-31.2"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="18.82,-34.45;53.165,-26.95"/>
+      </dia:attribute>
+      <dia:attribute name="meta">
+        <dia:composite type="dict"/>
+      </dia:attribute>
+      <dia:attribute name="orth_points">
+        <dia:point val="52.415,-31.2"/>
+        <dia:point val="52.415,-27"/>
+        <dia:point val="18.87,-27"/>
+        <dia:point val="18.87,-34.4"/>
+      </dia:attribute>
+      <dia:attribute name="orth_orient">
+        <dia:enum val="1"/>
+        <dia:enum val="0"/>
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="orth_autoroute">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="text_colour">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="line_colour">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:connections>
+        <dia:connection handle="0" to="O85" connection="6"/>
+        <dia:connection handle="1" to="O26" connection="6"/>
+      </dia:connections>
+    </dia:object>
+    <dia:object type="UML - Association" version="2" id="O87">
+      <dia:attribute name="name">
+        <dia:string>#creates file properties using &gt;#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="direction">
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="show_direction">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="assoc_type">
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="role_a">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="multipicity_a">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="visibility_a">
+        <dia:enum val="3"/>
+      </dia:attribute>
+      <dia:attribute name="show_arrow_a">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="role_b">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="multipicity_b">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="visibility_b">
+        <dia:enum val="3"/>
+      </dia:attribute>
+      <dia:attribute name="show_arrow_b">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="obj_pos">
+        <dia:point val="41,-38"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="40.25,-47.65;61.11,-36.4"/>
+      </dia:attribute>
+      <dia:attribute name="meta">
+        <dia:composite type="dict"/>
+      </dia:attribute>
+      <dia:attribute name="orth_points">
+        <dia:point val="41,-38"/>
+        <dia:point val="41,-43.15"/>
+        <dia:point val="60.91,-43.15"/>
+        <dia:point val="60.91,-47.6"/>
+      </dia:attribute>
+      <dia:attribute name="orth_orient">
+        <dia:enum val="1"/>
+        <dia:enum val="0"/>
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="orth_autoroute">
+        <dia:boolean val="true"/>
+      </dia:attribute>
+      <dia:attribute name="text_colour">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="line_colour">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:connections>
+        <dia:connection handle="0" to="O85" connection="0"/>
+        <dia:connection handle="1" to="O38" connection="7"/>
+      </dia:connections>
+    </dia:object>
+    <dia:object type="UML - Association" version="2" id="O88">
+      <dia:attribute name="name">
+        <dia:string>#creates file settings using &gt;#</dia:string>
+      </dia:attribute>
+      <dia:attribute name="direction">
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="show_direction">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="assoc_type">
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="role_a">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="multipicity_a">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="visibility_a">
+        <dia:enum val="3"/>
+      </dia:attribute>
+      <dia:attribute name="show_arrow_a">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="role_b">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="multipicity_b">
+        <dia:string>##</dia:string>
+      </dia:attribute>
+      <dia:attribute name="visibility_b">
+        <dia:enum val="3"/>
+      </dia:attribute>
+      <dia:attribute name="show_arrow_b">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="obj_pos">
+        <dia:point val="63.83,-38"/>
+      </dia:attribute>
+      <dia:attribute name="obj_bb">
+        <dia:rectangle val="63.08,-48.45;80.2,-36.4"/>
+      </dia:attribute>
+      <dia:attribute name="meta">
+        <dia:composite type="dict"/>
+      </dia:attribute>
+      <dia:attribute name="orth_points">
+        <dia:point val="63.83,-38"/>
+        <dia:point val="63.83,-43"/>
+        <dia:point val="80,-43"/>
+        <dia:point val="80,-48.4"/>
+      </dia:attribute>
+      <dia:attribute name="orth_orient">
+        <dia:enum val="1"/>
+        <dia:enum val="0"/>
+        <dia:enum val="1"/>
+      </dia:attribute>
+      <dia:attribute name="orth_autoroute">
+        <dia:boolean val="false"/>
+      </dia:attribute>
+      <dia:attribute name="text_colour">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:attribute name="line_colour">
+        <dia:color val="#000000"/>
+      </dia:attribute>
+      <dia:connections>
+        <dia:connection handle="0" to="O85" connection="2"/>
+        <dia:connection handle="1" to="O82" connection="5"/>
+      </dia:connections>
+    </dia:object>
   </dia:layer>
 </dia:diagram>
diff --git a/doc/agile/sprint_backlog_84.org b/doc/agile/sprint_backlog_84.org
index b615b3a..6e7fbd3 100644
--- a/doc/agile/sprint_backlog_84.org
+++ b/doc/agile/sprint_backlog_84.org
@@ -13,25 +13,25 @@
 ** Active
 
 #+begin: clocktable :maxlevel 3 :scope subtree :indent nil :emphasize nil :scope file :narrow 75 :formula %
-#+CAPTION: Clock summary at [2016-07-06 Wed 23:01]
-| <75>                                                                        |        |      |      |       |
-| Headline                                                                    | Time   |      |      |     % |
-|-----------------------------------------------------------------------------+--------+------+------+-------|
-| *Total time*                                                                | *9:56* |      |      | 100.0 |
-|-----------------------------------------------------------------------------+--------+------+------+-------|
-| Stories                                                                     | 9:56   |      |      | 100.0 |
-| Active                                                                      |        | 9:56 |      | 100.0 |
-| STARTED Sprint and product backlog grooming                                 |        |      | 0:55 |   9.2 |
-| STARTED Manual updates and blog posts                                       |        |      | 0:37 |   6.2 |
-| COMPLETED Try to fix tag build issues                                       |        |      | 1:28 |  14.8 |
-| COMPLETED Enable build type on travis                                       |        |      | 1:17 |  12.9 |
-| COMPLETED Add initial support for coveralls                                 |        |      | 0:29 |   4.9 |
-| COMPLETED Simplify local builds on emacs                                    |        |      | 0:20 |   3.4 |
-| COMPLETED Create a debug build for clang in travis                          |        |      | 0:10 |   1.7 |
-| COMPLETED Rename element settings back to aspect settings                   |        |      | 0:50 |   8.4 |
-| COMPLETED Rename bundle to element settings                                 |        |      | 0:27 |   4.5 |
-| COMPLETED Rename bundle in stitch to properties                             |        |      | 0:16 |   2.7 |
-| STARTED Refactor general settings                                           |        |      | 3:07 |  31.4 |
+#+CAPTION: Clock summary at [2016-07-07 Thu 21:02]
+| <75>                                                                        |         |       |      |       |
+| Headline                                                                    | Time    |       |      |     % |
+|-----------------------------------------------------------------------------+---------+-------+------+-------|
+| *Total time*                                                                | *11:27* |       |      | 100.0 |
+|-----------------------------------------------------------------------------+---------+-------+------+-------|
+| Stories                                                                     | 11:27   |       |      | 100.0 |
+| Active                                                                      |         | 11:27 |      | 100.0 |
+| STARTED Sprint and product backlog grooming                                 |         |       | 0:55 |   8.0 |
+| STARTED Manual updates and blog posts                                       |         |       | 0:37 |   5.4 |
+| COMPLETED Try to fix tag build issues                                       |         |       | 1:28 |  12.8 |
+| COMPLETED Enable build type on travis                                       |         |       | 1:17 |  11.2 |
+| COMPLETED Add initial support for coveralls                                 |         |       | 0:29 |   4.2 |
+| COMPLETED Simplify local builds on emacs                                    |         |       | 0:20 |   2.9 |
+| COMPLETED Create a debug build for clang in travis                          |         |       | 0:10 |   1.5 |
+| COMPLETED Rename element settings back to aspect settings                   |         |       | 0:50 |   7.3 |
+| COMPLETED Rename bundle to element settings                                 |         |       | 0:27 |   3.9 |
+| COMPLETED Rename bundle in stitch to properties                             |         |       | 0:16 |   2.3 |
+| STARTED Refactor general settings                                           |         |       | 4:38 |  40.5 |
 #+TBLFM: $5='(org-clock-time% @3$2 $2..$4);%.1f
 #+end:
 
@@ -253,6 +253,8 @@ In keeping with the properties/settings refactor, ensure stitch is
 using these terms with the same meanings as everyone else.
 
 *** STARTED Refactor general settings                                 :story:
+    CLOCK: [2016-07-07 Thu 20:18]--[2016-07-07 Thu 21:02] =>  0:44
+    CLOCK: [2016-07-07 Thu 07:53]--[2016-07-07 Thu 08:40] =>  0:47
     CLOCK: [2016-07-06 Wed 00:13]--[2016-07-06 Wed 00:20] =>  0:07
     CLOCK: [2016-07-06 Wed 00:06]--[2016-07-06 Wed 00:12] =>  0:06
     CLOCK: [2016-07-05 Tue 23:16]--[2016-07-06 Wed 00:05] =>  0:49
diff --git a/projects/formatters/include/dogen/formatters/types/all.hpp b/projects/formatters/include/dogen/formatters/types/all.hpp
index def2110..e5fcff7 100644
--- a/projects/formatters/include/dogen/formatters/types/all.hpp
+++ b/projects/formatters/include/dogen/formatters/types/all.hpp
@@ -62,6 +62,7 @@
 #include "dogen/formatters/types/cpp/namespace_formatter.hpp"
 #include "dogen/formatters/types/file_properties_factory.hpp"
 #include "dogen/formatters/types/modeline_group_hydrator.hpp"
+#include "dogen/formatters/types/file_properties_workflow.hpp"
 #include "dogen/formatters/types/cpp/boilerplate_formatter.hpp"
 #include "dogen/formatters/types/cpp/header_guard_formatter.hpp"
 #include "dogen/formatters/types/code_generation_marker_factory.hpp"
diff --git a/projects/formatters/include/dogen/formatters/types/file_properties_factory.hpp b/projects/formatters/include/dogen/formatters/types/file_properties_factory.hpp
index 12ed18e..38796c9 100644
--- a/projects/formatters/include/dogen/formatters/types/file_properties_factory.hpp
+++ b/projects/formatters/include/dogen/formatters/types/file_properties_factory.hpp
@@ -26,14 +26,12 @@
 #endif
 
 #include <string>
-#include <forward_list>
-#include <unordered_map>
 #include <boost/optional.hpp>
-#include "dogen/dynamic/types/object.hpp"
-#include "dogen/formatters/types/file_properties.hpp"
-#include "dogen/formatters/types/modeline_group.hpp"
 #include "dogen/formatters/types/licence.hpp"
 #include "dogen/formatters/types/repository.hpp"
+#include "dogen/formatters/types/modeline_group.hpp"
+#include "dogen/formatters/types/file_settings.hpp"
+#include "dogen/formatters/types/file_properties.hpp"
 
 namespace dogen {
 namespace formatters {
@@ -50,46 +48,27 @@ public:
     ~file_properties_factory() = default;
 
 public:
-    /**
-     * @brief Initialise a new file properties factory.
-     *
-     * @param rp where to look up reference data.
-     */
     explicit file_properties_factory(const repository& rp);
-
-    /**
-     * @brief Initialise a new file properties factory.
-     *
-     * @param rp where to look up reference data.
-     * @param fallback object to use to construct defaults, if any.
-     */
     file_properties_factory(const repository& rp,
-        const dynamic::object& fallback);
+        const file_settings& fallback);
 
 private:
     /**
-     * @brief Extracts the licence text from the dynamic object.
+     * @brief Obtains the licence text.
      */
     boost::optional<std::string>
-    extract_licence_text(const dynamic::object& o) const;
-
-    /**
-     * @brief Extracts the copyright notices from the dynamic object.
-     */
-    boost::optional<std::list<std::string>>
-    extract_copyright_notices(const dynamic::object& o) const;
+    get_licence_text(const file_settings& fs) const;
 
     /**
-     * @brief Extracts a licence from the dynamic object.
+     * @brief Obtains a licence.
      */
-    boost::optional<licence>
-    extract_licence(const dynamic::object& o) const;
+    boost::optional<licence> get_licence(const file_settings& fs) const;
 
     /**
-     * @brief Extracts the modeline group name from the dynamic object.
+     * @brief Obtains the modeline group name.
      */
     boost::optional<modeline_group>
-    extract_modeline_group(const dynamic::object& o) const;
+    get_modeline_group(const file_settings& fs) const;
 
     /**
      * @brief Returns the modeline for the supplied modeline name.
@@ -100,44 +79,35 @@ private:
         const modeline_group& mg) const;
 
     /**
-     * @brief Extracts a modeline the dynamic object.
+     * @brief Obtains the modeline given a modeline name.
      */
     boost::optional<modeline>
-    extract_modeline(const std::string& modeline_name,
-        const dynamic::object& o) const;
-
-    /**
-     * @brief Extracts a code generation marker the dynamic object.
-     */
-    boost::optional<std::string>
-    extract_marker(const dynamic::object& o) const;
+    get_modeline(const std::string& modeline_name,
+        const file_settings& fs) const;
 
     /**
-     * @brief Extracts the marker from the supplied dynamic object; if
-     * none is found, uses the default marker.
+     * @brief Obtains a code generation marker.
      */
-    std::string
-    extract_marker_or_default(const dynamic::object& o) const;
+    boost::optional<std::string> get_marker(const file_settings& fs) const;
 
     /**
-     * @brief Extracts the generate preamble field.
+     * @brief Obtains the marker from the file settings; if none is
+     * found, uses the default.
      */
-    boost::optional<bool>
-    extract_generate_preamble(const dynamic::object& o) const;
+    std::string get_marker_or_default(const file_settings& fs) const;
 
     /**
-     * @brief Extracts the generate preamble; if none is found, uses
+     * @brief Obtains the generate preamble; if none is found, uses
      * the default value.
      */
-    bool extract_generate_preamble_or_default(
-        const dynamic::object& o) const;
+    bool get_generate_preamble_or_default(const file_settings& fs) const;
 
 public:
     /**
-     * @brief Generates file properties from the dynamic object.
+     * @brief Generates the file properties.
      */
-    file_properties make(const std::string& modeline_name,
-        const dynamic::object& o) const;
+    file_properties
+    make(const std::string& modeline_name, const file_settings& fs) const;
 
 private:
     const repository& repository_;
diff --git a/projects/formatters/include/dogen/formatters/types/file_properties_workflow.hpp b/projects/formatters/include/dogen/formatters/types/file_properties_workflow.hpp
new file mode 100644
index 0000000..58d62d6
--- /dev/null
+++ b/projects/formatters/include/dogen/formatters/types/file_properties_workflow.hpp
@@ -0,0 +1,74 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_FORMATTERS_TYPES_FILE_PROPERTIES_WORKFLOW_HPP
+#define DOGEN_FORMATTERS_TYPES_FILE_PROPERTIES_WORKFLOW_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+#include "dogen/dynamic/types/object.hpp"
+#include "dogen/dynamic/types/repository.hpp"
+#include "dogen/formatters/types/repository.hpp"
+#include "dogen/formatters/types/file_properties.hpp"
+#include "dogen/formatters/types/file_settings_factory.hpp"
+#include "dogen/formatters/types/file_properties_factory.hpp"
+
+namespace dogen {
+namespace formatters {
+
+class file_properties_workflow {
+public:
+    /**
+     * @brief Initialise a new file properties workflow, without
+     * access to fallbacks. This constructor is used by stitch.
+     *
+     * @param rp where to look up reference data.
+     */
+    explicit file_properties_workflow(const dynamic::repository& drp,
+        const repository& rp);
+
+    /**
+     * @brief Initialise a new file properties workflow, with access
+     * to fallbacks. This constructor is used by quilt.cpp.
+     *
+     * @param rp where to look up reference data.
+     * @param fallback object to use to construct defaults, if any.
+     */
+    file_properties_workflow(const dynamic::repository& drp,
+        const repository& rp,
+        const dynamic::object& fallback);
+
+public:
+    /**
+     * @brief Generates file properties from the dynamic object.
+     */
+    file_properties execute(const std::string& modeline_name,
+        const dynamic::object& o) const;
+
+private:
+    const file_settings_factory settings_factory_;
+    const file_properties_factory properties_factory_;
+};
+
+} }
+
+#endif
diff --git a/projects/formatters/include/dogen/formatters/types/file_properties_workflow_fwd.hpp b/projects/formatters/include/dogen/formatters/types/file_properties_workflow_fwd.hpp
new file mode 100644
index 0000000..f0e445e
--- /dev/null
+++ b/projects/formatters/include/dogen/formatters/types/file_properties_workflow_fwd.hpp
@@ -0,0 +1,35 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#ifndef DOGEN_FORMATTERS_TYPES_FILE_PROPERTIES_WORKFLOW_FWD_HPP
+#define DOGEN_FORMATTERS_TYPES_FILE_PROPERTIES_WORKFLOW_FWD_HPP
+
+#if defined(_MSC_VER) && (_MSC_VER >= 1200)
+#pragma once
+#endif
+
+namespace dogen {
+namespace formatters {
+
+class file_properties_workflow;
+
+} }
+
+#endif
diff --git a/projects/formatters/include/dogen/formatters/types/file_settings_factory.hpp b/projects/formatters/include/dogen/formatters/types/file_settings_factory.hpp
index 242970f..2518b76 100644
--- a/projects/formatters/include/dogen/formatters/types/file_settings_factory.hpp
+++ b/projects/formatters/include/dogen/formatters/types/file_settings_factory.hpp
@@ -35,7 +35,7 @@ namespace dogen {
 namespace formatters {
 
 class file_settings_factory {
-private:
+public:
     explicit file_settings_factory(const dynamic::repository& rp);
 
 private:
diff --git a/projects/formatters/src/types/file_properties_factory.cpp b/projects/formatters/src/types/file_properties_factory.cpp
index c78222a..4de9951 100644
--- a/projects/formatters/src/types/file_properties_factory.cpp
+++ b/projects/formatters/src/types/file_properties_factory.cpp
@@ -20,9 +20,6 @@
  */
 #include <boost/throw_exception.hpp>
 #include "dogen/utility/log/logger.hpp"
-#include "dogen/dynamic/types/text.hpp"
-#include "dogen/dynamic/types/field_selector.hpp"
-#include "dogen/formatters/types/traits.hpp"
 #include "dogen/formatters/types/hydration_workflow.hpp"
 #include "dogen/formatters/types/code_generation_marker_factory.hpp"
 #include "dogen/formatters/types/building_error.hpp"
@@ -47,22 +44,20 @@ file_properties_factory::file_properties_factory(const repository& rp)
     : repository_(rp) { }
 
 file_properties_factory::file_properties_factory(const repository& rp,
-    const dynamic::object& fallback)
+    const file_settings& fallback)
     : repository_(rp),
-      default_licence_text_(extract_licence_text(fallback)),
-      default_copyright_notices_(extract_copyright_notices(fallback)),
-      default_modeline_group_(extract_modeline_group(fallback)),
-      default_marker_(extract_marker(fallback)),
-      default_generate_preamble_(extract_generate_preamble(fallback)) { }
-
-boost::optional<std::string> file_properties_factory::
-extract_licence_text(const dynamic::object& o) const {
-    using namespace dynamic;
-    const field_selector fs(o);
-    if (!fs.has_field(traits::licence_name()))
+      default_licence_text_(get_licence_text(fallback)),
+      default_copyright_notices_(fallback.copyright_notices()),
+      default_modeline_group_(get_modeline_group(fallback)),
+      default_marker_(get_marker(fallback)),
+      default_generate_preamble_(fallback.generate_preamble()) { }
+
+boost::optional<std::string>
+file_properties_factory::get_licence_text(const file_settings& fs) const {
+    if (fs.licence_name().empty())
         return boost::optional<std::string>();
 
-    const auto ln(fs.get_text_content(traits::licence_name()));
+    const auto ln(fs.licence_name());
     const auto i(repository_.licence_texts().find(ln));
     if (i == repository_.licence_texts().end()) {
         BOOST_LOG_SEV(lg, error) << licence_not_found << ln;
@@ -71,24 +66,14 @@ extract_licence_text(const dynamic::object& o) const {
     return i->second;
 }
 
-boost::optional<std::list<std::string> > file_properties_factory::
-extract_copyright_notices(const dynamic::object& o) const {
-    using namespace dynamic;
-    const field_selector fs(o);
+boost::optional<licence>
+file_properties_factory::get_licence(const file_settings& fs) const {
+    const auto overriden_licence_text(get_licence_text(fs));
+    const auto overriden_copyright_notices(fs.copyright_notices());
 
-    if (!fs.has_field(traits::copyright_notices()))
-        return boost::optional<std::list<std::string> >();
-
-    return fs.get_text_collection_content(traits::copyright_notices());
-}
-
-boost::optional<licence> file_properties_factory::
-extract_licence(const dynamic::object& o) const {
-    const auto overriden_licence_text(extract_licence_text(o));
-    const auto overriden_copyright_notices(extract_copyright_notices(o));
-
-    const bool has_licence(overriden_licence_text || default_licence_text_ ||
-        overriden_copyright_notices);
+    const bool has_text(overriden_licence_text || default_licence_text_);
+    const bool has_notices(overriden_copyright_notices.empty());
+    const bool has_licence(has_text || has_notices);
     if (!has_licence)
         return boost::optional<licence>();
 
@@ -98,22 +83,20 @@ extract_licence(const dynamic::object& o) const {
     else if (default_licence_text_)
         r.text(*default_licence_text_);
 
-    if (overriden_copyright_notices)
-        r.copyright_notices(*overriden_copyright_notices);
+    if (!overriden_copyright_notices.empty())
+        r.copyright_notices(overriden_copyright_notices);
     else if (default_copyright_notices_)
         r.copyright_notices(*default_copyright_notices_);
 
     return r;
 }
 
-boost::optional<modeline_group> file_properties_factory::
-extract_modeline_group(const dynamic::object& o) const {
-    using namespace dynamic;
-    const field_selector fs(o);
-    if (!fs.has_field(traits::modeline_group_name()))
+boost::optional<modeline_group>
+file_properties_factory::get_modeline_group(const file_settings& fs) const {
+    if (fs.modeline_group_name().empty())
         return boost::optional<modeline_group>();
 
-    const auto n(fs.get_text_content(traits::modeline_group_name()));
+    const auto n(fs.modeline_group_name());
     const auto i(repository_.modeline_groups().find(n));
     if (i == repository_.modeline_groups().end()) {
         BOOST_LOG_SEV(lg, error) << modeline_group_not_found << n;
@@ -133,11 +116,11 @@ modeline file_properties_factory::get_modeline_from_group(
     return i->second;
 }
 
-boost::optional<modeline> file_properties_factory::
-extract_modeline(const std::string& modeline_name,
-    const dynamic::object& o) const {
+boost::optional<modeline>
+file_properties_factory::get_modeline(const std::string& modeline_name,
+    const file_settings& fs) const {
 
-    const auto overridden_modeline_group(extract_modeline_group(o));
+    const auto overridden_modeline_group(get_modeline_group(fs));
     if (!overridden_modeline_group && !default_modeline_group_)
         return boost::optional<modeline>();
 
@@ -149,52 +132,36 @@ extract_modeline(const std::string& modeline_name,
     return get_modeline_from_group(modeline_name, *default_modeline_group_);
 }
 
-boost::optional<std::string> file_properties_factory::
-extract_marker(const dynamic::object& o) const {
-    using namespace dynamic;
-    using cgm = traits::code_generation_marker;
+boost::optional<std::string>
+file_properties_factory::get_marker(const file_settings& fs) const {
 
-    const field_selector fs(o);
-    if (!fs.has_field(cgm::message()))
+    const auto& msg(fs.marker_message());
+    if (msg.empty())
         return std::string();
 
-    const auto message(fs.get_text_content(cgm::message()));
-    if (message.empty())
-        return message;
-
-    const bool add_date_time(fs.get_boolean_content(cgm::add_date_time()));
-    const bool add_warning(fs.get_boolean_content(cgm::add_warning()));
-    code_generation_marker_factory f(add_date_time, add_warning, message);
-
+    const bool adt(*fs.marker_add_date_time());
+    const bool aw(*fs.marker_add_warning());
+    code_generation_marker_factory f(adt, aw, msg);
     return f.make();
 }
 
-std::string file_properties_factory::
-extract_marker_or_default(const dynamic::object& o) const {
-    const auto overridden_marker(extract_marker(o));
+std::string
+file_properties_factory::get_marker_or_default(const file_settings& fs) const {
+    const auto overridden_marker(get_marker(fs));
     if (overridden_marker)
         return *overridden_marker;
+
     if (default_marker_)
         return *default_marker_;
 
     return std::string();
 }
 
-boost::optional<bool> file_properties_factory::
-extract_generate_preamble(const dynamic::object& o) const {
-    using namespace dynamic;
-    const field_selector fs(o);
-    if (!fs.has_field(traits::generate_preamble()))
-        return boost::optional<bool>();
-
-    return fs.get_boolean_content(traits::generate_preamble());
-}
-
 bool file_properties_factory::
-extract_generate_preamble_or_default(const dynamic::object& o) const {
-    const auto overriden_generate_preamble(extract_generate_preamble(o));
+get_generate_preamble_or_default(const file_settings& fs) const {
+    const auto overriden_generate_preamble(fs.generate_preamble());
     if (!overriden_generate_preamble && !default_generate_preamble_)
-        return true; // random default.
+        return true; // FIXME: fairly random default.
 
     if (overriden_generate_preamble)
         return *overriden_generate_preamble;
@@ -204,12 +171,12 @@ extract_generate_preamble_or_default(const dynamic::object& o) const {
 
 file_properties
 file_properties_factory::make(const std::string& modeline_name,
-    const dynamic::object& o) const {
-    const auto modeline(extract_modeline(modeline_name, o));
-    const auto licence(extract_licence(o));
-    const auto marker(extract_marker_or_default(o));
+    const file_settings& fs) const {
+    const auto modeline(get_modeline(modeline_name, fs));
+    const auto licence(get_licence(fs));
+    const auto marker(get_marker_or_default(fs));
     const annotation a(modeline, licence, marker);
-    const bool gp(extract_generate_preamble_or_default(o));
+    const bool gp(get_generate_preamble_or_default(fs));
     return file_properties(gp, a);
 }
 
diff --git a/projects/formatters/src/types/file_properties_workflow.cpp b/projects/formatters/src/types/file_properties_workflow.cpp
new file mode 100644
index 0000000..7fdbd4b
--- /dev/null
+++ b/projects/formatters/src/types/file_properties_workflow.cpp
@@ -0,0 +1,43 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*-
+ *
+ * Copyright (C) 2012-2015 Marco Craveiro <marco.craveiro@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ *
+ */
+#include "dogen/formatters/types/file_properties_workflow.hpp"
+
+namespace dogen {
+namespace formatters {
+
+file_properties_workflow::file_properties_workflow(
+    const dynamic::repository& drp,
+    const repository& rp)
+    : settings_factory_(drp), properties_factory_(rp) { }
+
+file_properties_workflow::file_properties_workflow(
+    const dynamic::repository& drp, const repository& rp,
+    const dynamic::object& fallback)
+    : settings_factory_(drp),
+      properties_factory_(rp, *settings_factory_.make(fallback)) { }
+
+file_properties file_properties_workflow::execute(
+    const std::string& modeline_name, const dynamic::object& o) const {
+    const auto fs = settings_factory_.make(o);
+    return properties_factory_.make(modeline_name, fs);
+}
+
+} }
